{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "node_info = {}\n",
    "\n",
    "# There seem to be many types of sequence variants,\n",
    "# But the edges that link them are pretty unique, so will define based on\n",
    "# Those edges rather than a Q-ID\n",
    "node_info['Sequence Variant'] = {\n",
    "    'abbrev': 'V',\n",
    "    'id': None}\n",
    "\n",
    "node_info['Binding Site'] = {\n",
    "    'abbrev': 'BS',\n",
    "    'id': 'Q616005'}\n",
    "\n",
    "node_info['Active Site'] = {\n",
    "    'abbrev': 'AS',\n",
    "    'id': 'Q423026'}\n",
    "\n",
    "node_info['Super-Secondary Structure'] = {\n",
    "    'abbrev': 'SS',\n",
    "    'id': 'Q7644128'}\n",
    "\n",
    "node_info['Chemical Hazard'] = {\n",
    "    'abbrev': 'CH',\n",
    "    'id': 'Q21167512'}\n",
    "\n",
    "node_info['Molecular Function'] = {\n",
    "    'abbrev': 'MF',\n",
    "    'id': 'Q14860489'}\n",
    "    \n",
    "node_info['Pharmaceutical Product'] = {\n",
    "    'abbrev': 'PP',\n",
    "    'id': 'Q28885102'}\n",
    "\n",
    "node_info['Structural Motif'] = {\n",
    "    'abbrev': 'SM',\n",
    "    'id': 'Q3273544'}\n",
    "\n",
    "node_info['Protein Family'] = {\n",
    "    'abbrev': 'PF',\n",
    "    'id': 'Q417841'}\n",
    "\n",
    "node_info['Compound'] = {\n",
    "    'abbrev': 'C',\n",
    "    'id': 'Q11173'}\n",
    "\n",
    "node_info['Disease'] = {\n",
    "    'abbrev': 'D',\n",
    "    'id': 'Q12136'}\n",
    "\n",
    "## All proteins are encoded by genes, so qualify with 'gene encodes protein'\n",
    "node_info['Protein'] = {\n",
    "    'abbrev': 'P',\n",
    "    'id': 'Q8054',\n",
    "    'qualifiers': [{'s': 'Gene', 'p': 'P688', 'o': 'Protein', 'not': False},\n",
    "                   {'s': 'Gene', 'p': 'P703', 'o': 'Q15978631', 'not': False}]}\n",
    "\n",
    "## Genes will be defined as a human gene that does not encode a ANYTHING\n",
    "## None in Object means it only cares about P688.... (if you specifiy proteins, get memory erros...)\n",
    "node_info['Gene'] = {\n",
    "    'abbrev': 'G',\n",
    "    'id': 'Q7187',\n",
    "    'qualifiers': [{'s': 'Gene', 'p': 'P688', 'o': None, 'not': True}, \n",
    "                   {'s': 'Gene', 'p': 'P703', 'o': 'Q15978631', 'not': False}]}\n",
    "    \n",
    "node_info['Protein Domain'] = {\n",
    "    'abbrev': 'PD',\n",
    "    'id': 'Q898273'}\n",
    "    \n",
    "node_info['Cellular Component'] = {\n",
    "    'abbrev': 'CC',\n",
    "    'id': 'Q5058355'}\n",
    "    \n",
    "node_info['Biological Pathway'] = {\n",
    "    'abbrev': 'PW',\n",
    "    'id': 'Q4915012'}\n",
    "\n",
    "node_info['Biological Process'] = {\n",
    "    'abbrev': 'BP',\n",
    "    'id': 'Q2996394'}\n",
    "\n",
    "# Don't want symptoms that could also be diseases....\n",
    "# Get a lot of overlapping types espectially on chemcial hazard edges\n",
    "node_info['Symptom'] = {\n",
    "    'abbrev': 'S',\n",
    "    'id': 'Q169872',\n",
    "    'qualifiers': [{'s': 'Symptom', 'p': 'P31', 'o': 'Q12136', 'not': True},\n",
    "                   {'s': 'Symptom', 'p': 'P279', 'o': 'Q12136', 'not': True}]}\n",
    "    \n",
    "node_info['Medical Specialty'] = {\n",
    "    'abbrev': 'MS',\n",
    "    'id': 'Q930752'}\n",
    "\n",
    "# Chemical Role doesn't exist, but is an important type....\n",
    "# Again defining based on the edge\n",
    "node_info['Chemical Role'] = {\n",
    "    'abbrev': 'CR',\n",
    "    'id': None}\n",
    "    \n",
    "node_info['Anatomical Structure'] = {\n",
    "    'abbrev': 'A',\n",
    "    'id': 'Q4936952'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "edge_info = {}\n",
    "\n",
    "edge_info['BIOLOGICAL_PROCESS'] = {\n",
    "    'abbrev': 'bp',\n",
    "    'id': 'P682'}\n",
    "\n",
    "edge_info['VARIANT_OF'] = {\n",
    "    'abbrev': 'v',\n",
    "    'id': 'P3433'}\n",
    "\n",
    "edge_info['CELL_COMPONENT'] = {\n",
    "    'abbrev': 'cc',\n",
    "    'id': 'P681'}\n",
    "\n",
    "edge_info['TREATS'] = {\n",
    "    'abbrev': 't',\n",
    "    'id': 'P2175',\n",
    "    'rev_id': 'P2176'}\n",
    "\n",
    "edge_info['GENETIC_ASSOCIATION'] = {\n",
    "    'abbrev': 'a',\n",
    "    'id': 'P2293'}\n",
    "\n",
    "edge_info['HAS_ACTIVE_INGREDIENT'] = {\n",
    "    'abbrev': 'ai',\n",
    "    'id': 'P3781'}\n",
    "\n",
    "edge_info['HAS_CAUSE'] = {\n",
    "    'abbrev': 'hc',\n",
    "    'id': 'P828'}\n",
    "\n",
    "edge_info['HAS_PART'] = {\n",
    "    'abbrev': 'hp',\n",
    "    'id': 'P527'}\n",
    "\n",
    "edge_info['MOLECULAR_FUNCTION'] = {\n",
    "    'abbrev': 'mf',\n",
    "    'id': 'P680'}\n",
    "\n",
    "edge_info['NEGATIVE_DIAGNOSTIC_PREDICTOR'] = {\n",
    "    'abbrev': 'nd',\n",
    "    'id': 'P3357'}\n",
    "\n",
    "edge_info['NEGATIVE_THERAPEUTIC_PREDICTOR'] = {\n",
    "    'abbrev': 'nt',\n",
    "    'id': 'P3355'}\n",
    "\n",
    "edge_info['PHYSICALLY_INTERACTS_WITH'] = {\n",
    "    'abbrev': 'iw',\n",
    "    'id': 'P129'}\n",
    "\n",
    "edge_info['POSITIVE_DIAGNOSTIC_PREDICTOR'] = {\n",
    "    'abbrev': 'pd',\n",
    "    'id': 'P3356'}\n",
    "\n",
    "edge_info['POSITIVE_THERAPEUTIC_PREDICTOR'] = {\n",
    "    'abbrev': 'pt',\n",
    "    'id': 'P3354'}\n",
    "\n",
    "edge_info['SUBCLASS_OF'] = {\n",
    "    'abbrev': 's',\n",
    "    'id': 'P279'}\n",
    "\n",
    "edge_info['SYMPTOMS'] = {\n",
    "    'abbrev': 'sy',\n",
    "    'id': 'P780'}\n",
    "\n",
    "edge_info['MEDICAL_SPECIALTY'] = {\n",
    "    'abbrev': 'ms',\n",
    "    'id': 'P1995'}\n",
    "\n",
    "edge_info['SIGNIFICANT_DRUG_INTERACTION'] = {\n",
    "    'abbrev': 'di',\n",
    "    'id': 'P769'}\n",
    "\n",
    "edge_info['AFFLICTS'] = {\n",
    "    'abbrev': 'af',\n",
    "    'id': 'P689'}\n",
    "\n",
    "edge_info['HAS_ROLE'] = {\n",
    "    'abbrev': 'r',\n",
    "    'id': 'P2868'}\n",
    "\n",
    "edge_info['THERAPEUTIC_AREA'] = {\n",
    "    'abbrev': 'ta',\n",
    "    'id': 'P4044'}\n",
    "\n",
    "edge_info['ANATOMICAL_LOCATION'] = {\n",
    "    'abbrev': 'l',\n",
    "    'id': 'P927'}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "node_abbreviations = {k: v['abbrev'] for k, v in node_info.items()}\n",
    "edge_abbreviations = {k: v['abbrev'] for k, v in edge_info.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "node_abv_to_full = {v: k for k, v in node_abbreviations.items()}\n",
    "edge_abv_to_full = {v: k for k, v in edge_abbreviations.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate queries for the edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../../hetnet-ml/src/')\n",
    "import graph_tools as gt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def generate_instance_tag(qname, n_id):\n",
    "    return \"\\n    {n} wdt:P31|wdt:P279 wd:{n_id} .\".format(n=qname, n_id=n_id)\n",
    "\n",
    "def to_query_name(name):\n",
    "    return '?'+name.lower().replace(' ', '_').replace('-', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_query_from_abbrev(abbrev, target=None):\n",
    "    # Split the edge abbreviation in to subject, predicate, object\n",
    "    s_abv, p_abv, o_abv = gt.parse_edge_abbrev(abbrev)\n",
    "    \n",
    "    # Get the full versions of names\n",
    "    s_name = node_abv_to_full[s_abv]\n",
    "    p_name = edge_abv_to_full[p_abv]\n",
    "    o_name = node_abv_to_full[o_abv]\n",
    "    \n",
    "    # Sanitize names for query purposes\n",
    "    s_qname = to_query_name(s_name)\n",
    "    o_qname = to_query_name(o_name)\n",
    "    \n",
    "    # Need to differentiate start and end if self referential edge\n",
    "    if s_qname == o_qname:\n",
    "        s_qname += '1'\n",
    "        o_qname += '2'\n",
    "    \n",
    "    # Get the proper return types\n",
    "    query_text = \"SELECT DISTINCT {s} {s}Label {o} {o}Label\".format(s=s_qname, o=o_qname)\n",
    "    query_text += \"\\nWHERE {\"\n",
    "    \n",
    "    return_types = {o_name}\n",
    "    \n",
    "    # Build the appropriate info to get the correct subject and object node-types\n",
    "    # Use set so not repeated in cases of self-referential edges\n",
    "    for name, qname in [(s_name, s_qname), (o_name, o_qname)]:\n",
    "        if node_info[name]['id']:\n",
    "            query_text += '\\n\\n    # Initial typing for {}'.format(name)\n",
    "            query_text += generate_instance_tag(qname, node_info[name]['id'])\n",
    "            return_types.update({name})\n",
    "    \n",
    "        # Get any qualifiers for subjects and objects\n",
    "        for qual in node_info[name].get('qualifiers', []):\n",
    "            query_text += '\\n    # Qualifier for {}'.format(name)\n",
    "            query_text += generate_qualifier(qual, return_types=return_types)\n",
    "            # Ensure there isn't the same nodetype definition multiple times\n",
    "            # E.G. ?gene wdt:P31|wdt:P279 wd:wd:Q7187 # ?gene subclass/instace of Gene\n",
    "            return_types.update({qual['s'], qual['o']})\n",
    "                \n",
    "    # Allows for changing of the target of an edge... for example if we want \n",
    "    # Sequnce Variant - variant of - Protein... SV's only go to Genes, so we can use \n",
    "    # SV - v - Gene - enc - Pro to get the relationship... \n",
    "    # Qualifiers in the protein definito handle the Gene-enc-Pro relationship, but the Target\n",
    "    # variable handles the fact that SVs point to genes\n",
    "    if target:\n",
    "        o_qname = to_query_name(target)\n",
    "    \n",
    "    # The actual edge we're interested in\n",
    "    query_text += \"\\n\\n    # Edge of interest {s} {p} {o}\".format(s=s_name, p=p_name, o=o_name)\n",
    "    \n",
    "    # Some edges are bi-directonal e.g. drug-used-for-treatment and medical-condition-treated\n",
    "    # Get edges in eitehr dircetion\n",
    "    rev_id = edge_info[p_name].get('rev_id', None)\n",
    "    if rev_id is not None:\n",
    "        query_text += \"\\n    {{ {s} wdt:{p} {o} }}\".format(s=s_qname, p=edge_info[p_name]['id'], o=o_qname)\n",
    "        query_text += \"\\n    UNION {{ {s} wdt:{p} {o} }}\".format(s=o_qname, p=rev_id, o=s_qname)\n",
    "    else:\n",
    "        query_text += \"\\n    {s} wdt:{p} {o} \".format(s=s_qname, p=edge_info[p_name]['id'], o=o_qname)\n",
    "\n",
    "    \n",
    "    # Make sure labels return in english\n",
    "    query_text += '\\n\\n    SERVICE wikibase:label { bd:serviceParam wikibase:language \"[AUTO_LANGAGE],en\" }\\n}'\n",
    "\n",
    "    return query_text\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def generate_qualifier(qual, return_types=None):\n",
    "    node_q_names = {}\n",
    "    \n",
    "    for node in ['s', 'o']:\n",
    "        node_name = qual[node]\n",
    "        if node_name in node_info.keys():\n",
    "            node_q_names[node] = to_query_name(node_name)\n",
    "        elif node_name is None:\n",
    "            node_q_names[node] = '?na'\n",
    "        else:    \n",
    "            node_q_names[node] = 'wd:'+node_name\n",
    "\n",
    "    qual_out = ''\n",
    "    \n",
    "    # if we have true node-types, we need to ensure they're used\n",
    "    for semtype, qname in node_q_names.items():\n",
    "        if not qname.startswith('wd:') and qual[semtype] not in return_types and qname is not '?na':\n",
    "            qual_out += generate_instance_tag(qname, node_info[qual[semtype]]['id'])        \n",
    "    \n",
    "    # Negative qualifiers need to filter out edges\n",
    "    if qual['not']:\n",
    "        qual_out += '\\n    FILTER NOT EXISTS {{ {s} wdt:{p} {o} }} .'.format(s=node_q_names['s'], \n",
    "                                                                            p=qual['p'], o=node_q_names['o'])\n",
    "    # Otherwise add the edges\n",
    "    else:\n",
    "        qual_out += '\\n    {s} wdt:{p} {o} .'.format(s=node_q_names['s'], \n",
    "                                                    p=qual['p'], o=node_q_names['o'])\n",
    "    \n",
    "    return qual_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "hetnet_edges = [\n",
    "    {'abbrev': 'CdiC'},\n",
    "    {'abbrev': 'CtD'},\n",
    "    {'abbrev': 'PPaiC'},\n",
    "    {'abbrev': 'CHhcC'},\n",
    "    {'abbrev': 'PWhpC'},\n",
    "    {'abbrev': 'PiwC'},\n",
    "    {'abbrev': 'VntC'},\n",
    "    {'abbrev': 'VptC'},\n",
    "    {'abbrev': 'DaP', 'target': 'Gene'},\n",
    "    {'abbrev': 'DaG'},\n",
    "    {'abbrev': 'DsyS'},\n",
    "    {'abbrev': 'DmsMS'},\n",
    "    {'abbrev': 'CHsyS'},\n",
    "    {'abbrev': 'CHsyD'},\n",
    "    {'abbrev': 'VndD'},\n",
    "    {'abbrev': 'VpdD'},\n",
    "    {'abbrev': 'VvP', 'target': 'Gene'},\n",
    "    {'abbrev': 'VvG'},\n",
    "    {'abbrev': 'PWhpP', 'target': 'Gene'},\n",
    "    {'abbrev': 'PWhpG'},\n",
    "    {'abbrev': 'PccCC'},\n",
    "    {'abbrev': 'PbpBP'},\n",
    "    {'abbrev': 'PmfMF'},\n",
    "    {'abbrev': 'PhpPD'},\n",
    "    {'abbrev': 'PhpSS'},\n",
    "    {'abbrev': 'PsPF'},\n",
    "    {'abbrev': 'PhpBS'},\n",
    "    {'abbrev': 'PhpAS'},\n",
    "    {'abbrev': 'PhpSM'},\n",
    "    {'abbrev': 'PPtaD'},\n",
    "    {'abbrev': 'CrCR'},\n",
    "    {'abbrev': 'DlA'},\n",
    "    {'abbrev': 'CHafA'},\n",
    "    {'abbrev': 'CtCH'}\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "queries = [build_query_from_abbrev(**g_edge) for g_edge in hetnet_edges]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run the queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import functools\n",
    "from wikidataintegrator.wdi_core import WDItemEngine\n",
    "from tqdm import tqdm\n",
    "\n",
    "endpoint='http://avalanche.scripps.edu:9999/bigdata/sparql'\n",
    "\n",
    "def parse_result_uris(result):\n",
    "    for c in result:\n",
    "        if 'Label' not in c:\n",
    "            idx = result[c].str.startswith('http://www.wikidata.org/entity')\n",
    "            result[c][idx] = result[c][idx].str.split('/', expand=True).iloc[:,-1]\n",
    "    return result.drop_duplicates()\n",
    "\n",
    "query_func = functools.partial(WDItemEngine.execute_sparql_query, endpoint=endpoint, as_dataframe=True)\n",
    "\n",
    "def execute_sparql_query(query_text):\n",
    "    # Enforce the proper column order\n",
    "    col_order = query_text.split('\\n')[0].split(' ?')[1:]\n",
    "    return parse_result_uris(query_func(query_text))[col_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 34/34 [03:33<00:00,  6.27s/it]\n"
     ]
    }
   ],
   "source": [
    "res = {e['abbrev']: execute_sparql_query(q) for e, q in tqdm(zip(hetnet_edges, queries), total=len(hetnet_edges))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'460,776 raw edges'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many total edges retrieved?\n",
    "'{:,} raw edges'.format(sum(len(v) for v in res.values()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Two potnetial issues to look into...\n",
    "\n",
    "1. Self-referential edges...  \n",
    "    These edges may contain `node1 - node2` edges as well as `node2 - node1` edges\n",
    "    \n",
    "2. Multi or miss-typed nodes...  \n",
    "    Some node IDs may show up as multiple types... especially chemical roles and chemical compounds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_ids = dict()\n",
    "id_to_name = dict()\n",
    "self_ref = set()\n",
    "\n",
    "for e, r in res.items():\n",
    "    n_types = [c for c in r.columns if not c.endswith('Label')]\n",
    "    \n",
    "    for nt in n_types:\n",
    "        if nt.endswith('1') or nt.endswith('2'):\n",
    "            node_type = nt[:-1]\n",
    "            self_ref.add(e)\n",
    "        else:\n",
    "            node_type = nt\n",
    "    \n",
    "        if node_type in node_ids:\n",
    "            node_ids[node_type].update(set(r[nt]))\n",
    "        else:\n",
    "            node_ids[node_type] = set(r[nt])\n",
    "    \n",
    "        id_to_name.update(r.set_index(nt)[nt+'Label'].to_dict())\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fix the Self-Ref edges where forward and backward edges may be duplicated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(res.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00, 85.04it/s]\n"
     ]
    }
   ],
   "source": [
    "fixed = dict()\n",
    "\n",
    "for kind in tqdm(self_ref):\n",
    "    \n",
    "    # no need to worry about forward vs reverse in directed edges\n",
    "    if '>' in kind or '<' in kind:\n",
    "        continue\n",
    "    \n",
    "    # Only look at 1 kind of edge at a time\n",
    "    this_edges = res[kind]\n",
    "    col_names = this_edges.columns\n",
    "    \n",
    "    edge_ids = set()\n",
    "    \n",
    "    for row in this_edges.itertuples():\n",
    "        # Grab the edge ID, sorting, so lowest ID first:\n",
    "        #     If both 'Q00001 -- Q00002' and 'Q00002 -- Q00001' exist, effectively standarizes to \n",
    "        #     'Q00001 -- Q00002'\n",
    "        edge_id = tuple(sorted([row[1], row[3]]))\n",
    "        edge_ids.add(edge_id)\n",
    "\n",
    "    start_ids = []\n",
    "    start_names = []\n",
    "    end_ids = []\n",
    "    end_names = []\n",
    "\n",
    "        \n",
    "    for edge_id in edge_ids:\n",
    "        start_ids.append(edge_id[0])\n",
    "        start_names.append(id_to_name[edge_id[0]])\n",
    "        end_ids.append(edge_id[1])\n",
    "        end_names.append(id_to_name[edge_id[1]])\n",
    "    \n",
    "    fixed[kind] = pd.DataFrame({col_names[0]: start_ids, col_names[1]: start_names, col_names[2]: end_ids, col_names[3]: end_names})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge direction standardization results for CdiC\n",
      "Before: 1,807\tAfter: 1,105\n"
     ]
    }
   ],
   "source": [
    "# Check an arbitrary edge that was fixed\n",
    "k = next(iter(fixed.keys()))\n",
    "print('Edge direction standardization results for {}'.format(k))\n",
    "print('Before: {:,}\\tAfter: {:,}'.format(len(res[k]), len(fixed[k])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'66,677 nodes (possibly some duplicats across type)'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# numver of unique node ids\n",
    "'{:,} nodes (possibly some duplicats across type)'.format(sum(len(v) for v in node_ids.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "compound : 8089\n",
      "disease : 6207\n",
      "pharmaceutical_product : 2301\n",
      "chemical_hazard : 689\n",
      "biological_pathway : 563\n",
      "protein : 22742\n",
      "sequence_variant : 1758\n",
      "gene : 215\n",
      "symptom : 253\n",
      "medical_specialty : 69\n",
      "cellular_component : 1729\n",
      "biological_process : 12126\n",
      "molecular_function : 4267\n",
      "protein_domain : 4296\n",
      "supersecondary_structure : 455\n",
      "protein_family : 16\n",
      "binding_site : 55\n",
      "active_site : 89\n",
      "structural_motif : 108\n",
      "chemical_role : 494\n",
      "anatomical_structure : 156\n"
     ]
    }
   ],
   "source": [
    "for k, v in node_ids.items():\n",
    "    print('{} : {}'.format(k, len(v)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'66,673 unique nodes ids'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total = set()\n",
    "for v in node_ids.values():\n",
    "    total = total.union(v) \n",
    "'{:,} unique nodes ids'.format(len(total))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'4 total nodes with multiple types'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'{:,} total nodes with multiple types'.format(sum(len(v) for v in node_ids.values()) - len(total))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It will be easier to address these node issues once we've got the networks formatted into the hetnet standard\n",
    "\n",
    "## Format Nodes into Hetnet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes = []\n",
    "for k, v in node_ids.items():\n",
    "    curr_nodes = pd.DataFrame({'id': list(v), 'label': len(v)*[k]})\n",
    "    curr_nodes['name'] = curr_nodes['id'].map(id_to_name)\n",
    "    nodes.append(curr_nodes)\n",
    "nodes = pd.concat(nodes).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1413</th>\n",
       "      <td>Q127060</td>\n",
       "      <td>compound</td>\n",
       "      <td>folic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66431</th>\n",
       "      <td>Q127060</td>\n",
       "      <td>chemical_role</td>\n",
       "      <td>folic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8583</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>disease</td>\n",
       "      <td>mercury poisoning</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16979</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>chemical_hazard</td>\n",
       "      <td>mercury poisoning</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2665</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>compound</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32949</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>protein</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14077</th>\n",
       "      <td>Q7860879</td>\n",
       "      <td>disease</td>\n",
       "      <td>type IV hypersensitivity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50473</th>\n",
       "      <td>Q7860879</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>type IV hypersensitivity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             id               label                           name\n",
       "1413    Q127060            compound                     folic acid\n",
       "66431   Q127060       chemical_role                     folic acid\n",
       "8583    Q408089             disease              mercury poisoning\n",
       "16979   Q408089     chemical_hazard              mercury poisoning\n",
       "2665    Q414964            compound  Vasoactive intestinal peptide\n",
       "32949   Q414964             protein  Vasoactive intestinal peptide\n",
       "14077  Q7860879             disease       type IV hypersensitivity\n",
       "50473  Q7860879  biological_process       type IV hypersensitivity"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes[nodes.duplicated(keep=False, subset=['id'])].sort_values('id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "label_map = {to_query_name(k)[1:]: k for k in node_info.keys()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "nodes['label'] = nodes['label'].map(label_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q27107534</td>\n",
       "      <td>Compound</td>\n",
       "      <td>nitroblue tetrazolium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q27102886</td>\n",
       "      <td>Compound</td>\n",
       "      <td>(S)-3-aminoisobutyrate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q4836835</td>\n",
       "      <td>Compound</td>\n",
       "      <td>BW723C86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q2112916</td>\n",
       "      <td>Compound</td>\n",
       "      <td>propachlor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q411092</td>\n",
       "      <td>Compound</td>\n",
       "      <td>pyrophosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q6823517</td>\n",
       "      <td>Compound</td>\n",
       "      <td>metergoline</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Q1187179</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Denileukin diftitox</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q424390</td>\n",
       "      <td>Compound</td>\n",
       "      <td>cyclopentadiene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q413598</td>\n",
       "      <td>Compound</td>\n",
       "      <td>prilocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q27265984</td>\n",
       "      <td>Compound</td>\n",
       "      <td>carbizocaine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          id     label                    name\n",
       "0  Q27107534  Compound   nitroblue tetrazolium\n",
       "1  Q27102886  Compound  (S)-3-aminoisobutyrate\n",
       "2   Q4836835  Compound                BW723C86\n",
       "3   Q2112916  Compound              propachlor\n",
       "4    Q411092  Compound     pyrophosphoric acid\n",
       "5   Q6823517  Compound             metergoline\n",
       "6   Q1187179  Compound     Denileukin diftitox\n",
       "7    Q424390  Compound         cyclopentadiene\n",
       "8    Q413598  Compound              prilocaine\n",
       "9  Q27265984  Compound            carbizocaine"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format Edges into Hetnet\n",
    "\n",
    "The original design feature was to merege `Genes` and `Proetins` acrsos the `ENCODES` edge to create one metanode `Gene` that essentially means Gene or Gene Product.  We will start that process when building the edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges = []\n",
    "\n",
    "for k, v in res.items():\n",
    "    if k in fixed.keys():\n",
    "        v = fixed[k]\n",
    "    \n",
    "    \n",
    "    col_names = v.columns\n",
    "    keep_cols = [c for c in col_names if not c.endswith('Label')]\n",
    "    col_name_map = {keep_cols[0]: 'start_id', keep_cols[1]: 'end_id'}\n",
    "    \n",
    "    # Replace Proteins with Genes, to merge the protein and gene metanodes\n",
    "    parsed_edge = gt.parse_edge_abbrev(k)\n",
    "    if 'P' in parsed_edge:\n",
    "        idx = parsed_edge.index('P')\n",
    "        parsed_edge = list(parsed_edge)\n",
    "        parsed_edge[idx] = 'G'\n",
    "        k = ''.join(parsed_edge)\n",
    "    \n",
    "    v = v[keep_cols].rename(columns=col_name_map)\n",
    "    v['type'] = edge_abv_to_full[parsed_edge[1]] + '_' + k\n",
    "    \n",
    "    edges.append(v)\n",
    "edges = pd.concat(edges).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BIOLOGICAL_PROCESS_GbpBP               161545\n",
       "CELL_COMPONENT_GccCC                    99926\n",
       "MOLECULAR_FUNCTION_GmfMF                87111\n",
       "HAS_PART_GhpPD                          30490\n",
       "HAS_PART_PWhpG                          24876\n",
       "HAS_ROLE_CrCR                            9931\n",
       "MEDICAL_SPECIALTY_DmsMS                  6533\n",
       "TREATS_CtD                               5217\n",
       "HAS_PART_GhpSS                           4130\n",
       "PHYSICALLY_INTERACTS_WITH_GiwC           3645\n",
       "SYMPTOMS_CHsyS                           3450\n",
       "HAS_PART_PWhpC                           3384\n",
       "GENETIC_ASSOCIATION_DaG                  3043\n",
       "VARIANT_OF_VvG                           2528\n",
       "HAS_PART_GhpSM                           2238\n",
       "HAS_ACTIVE_INGREDIENT_PPaiC              2162\n",
       "SYMPTOMS_CHsyD                           1481\n",
       "AFFLICTS_CHafA                           1170\n",
       "THERAPEUTIC_AREA_PPtaD                   1124\n",
       "SIGNIFICANT_DRUG_INTERACTION_CdiC        1105\n",
       "HAS_PART_GhpAS                           1046\n",
       "HAS_PART_GhpBS                            852\n",
       "POSITIVE_THERAPEUTIC_PREDICTOR_VptC       721\n",
       "ANATOMICAL_LOCATION_DlA                   704\n",
       "HAS_CAUSE_CHhcC                           597\n",
       "NEGATIVE_THERAPEUTIC_PREDICTOR_VntC       541\n",
       "SYMPTOMS_DsyS                             372\n",
       "POSITIVE_DIAGNOSTIC_PREDICTOR_VpdD         78\n",
       "TREATS_CtCH                                47\n",
       "SUBCLASS_OF_GsPF                           21\n",
       "NEGATIVE_DIAGNOSTIC_PREDICTOR_VndD          6\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges['type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# merge the genes and proteins in the nodes file\n",
    "idx = nodes.query('label == \"Protein\"').index\n",
    "nodes.loc[idx, 'label'] = 'Gene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SELECT DISTINCT ?sequence_variant ?sequence_variantLabel ?disease ?diseaseLabel\n",
      "WHERE {\n",
      "\n",
      "    # Initial typing for Disease\n",
      "    ?disease wdt:P31|wdt:P279 wd:Q12136 .\n",
      "\n",
      "    # Edge of interest Sequence Variant NEGATIVE_DIAGNOSTIC_PREDICTOR Disease\n",
      "    ?sequence_variant wdt:P3357 ?disease \n",
      "\n",
      "    SERVICE wikibase:label { bd:serviceParam wikibase:language \"[AUTO_LANGAGE],en\" }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(build_query_from_abbrev('VndD'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene                         22957\n",
       "Biological Process           12126\n",
       "Compound                      8089\n",
       "Disease                       6207\n",
       "Protein Domain                4296\n",
       "Molecular Function            4267\n",
       "Pharmaceutical Product        2301\n",
       "Sequence Variant              1758\n",
       "Cellular Component            1729\n",
       "Chemical Hazard                689\n",
       "Biological Pathway             563\n",
       "Chemical Role                  494\n",
       "Super-Secondary Structure      455\n",
       "Symptom                        253\n",
       "Anatomical Structure           156\n",
       "Structural Motif               108\n",
       "Active Site                     89\n",
       "Medical Specialty               69\n",
       "Binding Site                    55\n",
       "Protein Family                  16\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes['label'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now Back to problem 2, nodes that are duplicated across two different Node Types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "duplicated_nodes = nodes[nodes.duplicated(keep=False, subset=['id'])]['id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Q127060', 'Q414964', 'Q408089', 'Q7860879'], dtype=object)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicated_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "combo = gt.combine_nodes_and_edges(nodes, edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>start_name</th>\n",
       "      <th>end_name</th>\n",
       "      <th>start_label</th>\n",
       "      <th>end_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q244408</td>\n",
       "      <td>Q407431</td>\n",
       "      <td>SIGNIFICANT_DRUG_INTERACTION_CdiC</td>\n",
       "      <td>diclofenac</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>Q408801</td>\n",
       "      <td>Q170990</td>\n",
       "      <td>TREATS_CtD</td>\n",
       "      <td>celecoxib</td>\n",
       "      <td>arthritis</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6322</th>\n",
       "      <td>Q47521989</td>\n",
       "      <td>Q409251</td>\n",
       "      <td>HAS_ACTIVE_INGREDIENT_PPaiC</td>\n",
       "      <td>Phoslo</td>\n",
       "      <td>calcium acetate</td>\n",
       "      <td>Pharmaceutical Product</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8484</th>\n",
       "      <td>Q21175000</td>\n",
       "      <td>Q417316</td>\n",
       "      <td>HAS_CAUSE_CHhcC</td>\n",
       "      <td>nitroethane exposure</td>\n",
       "      <td>nitroethane</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9081</th>\n",
       "      <td>Q44054558</td>\n",
       "      <td>Q27104149</td>\n",
       "      <td>HAS_PART_PWhpC</td>\n",
       "      <td>Glycolysis and Gluconeogenesis</td>\n",
       "      <td>D-malic acid</td>\n",
       "      <td>Biological Pathway</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12465</th>\n",
       "      <td>Q4847907</td>\n",
       "      <td>Q411917</td>\n",
       "      <td>PHYSICALLY_INTERACTS_WITH_GiwC</td>\n",
       "      <td>Cholinergic receptor muscarinic 4</td>\n",
       "      <td>glycopyrrolate</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16110</th>\n",
       "      <td>Q28546459</td>\n",
       "      <td>Q18936</td>\n",
       "      <td>NEGATIVE_THERAPEUTIC_PREDICTOR_VntC</td>\n",
       "      <td>RB1 OVEREXPRESSION</td>\n",
       "      <td>doxorubicin</td>\n",
       "      <td>Sequence Variant</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16651</th>\n",
       "      <td>Q29920046</td>\n",
       "      <td>Q27077182</td>\n",
       "      <td>POSITIVE_THERAPEUTIC_PREDICTOR_VptC</td>\n",
       "      <td>IDH2 MUTATION</td>\n",
       "      <td>Enasidenib</td>\n",
       "      <td>Sequence Variant</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17372</th>\n",
       "      <td>Q844935</td>\n",
       "      <td>Q4650352</td>\n",
       "      <td>GENETIC_ASSOCIATION_DaG</td>\n",
       "      <td>coronary artery disease</td>\n",
       "      <td>ABO, alpha 1-3-N-acetylgalactosaminyltransfera...</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20415</th>\n",
       "      <td>Q4454701</td>\n",
       "      <td>Q186889</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>Taeniasis saginata</td>\n",
       "      <td>nausea</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20787</th>\n",
       "      <td>Q373822</td>\n",
       "      <td>Q7867</td>\n",
       "      <td>MEDICAL_SPECIALTY_DmsMS</td>\n",
       "      <td>eating disorder</td>\n",
       "      <td>psychiatry</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Medical Specialty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27320</th>\n",
       "      <td>Q7030575</td>\n",
       "      <td>Q86</td>\n",
       "      <td>SYMPTOMS_CHsyS</td>\n",
       "      <td>nicotine poisoning</td>\n",
       "      <td>headache</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30770</th>\n",
       "      <td>Q21167640</td>\n",
       "      <td>Q1799146</td>\n",
       "      <td>SYMPTOMS_CHsyD</td>\n",
       "      <td>acrylic acid exposure</td>\n",
       "      <td>irritant dermatitis</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32251</th>\n",
       "      <td>Q27908043</td>\n",
       "      <td>Q264118</td>\n",
       "      <td>NEGATIVE_DIAGNOSTIC_PREDICTOR_VndD</td>\n",
       "      <td>DNMT3A R882</td>\n",
       "      <td>acute myeloid leukemia</td>\n",
       "      <td>Sequence Variant</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32257</th>\n",
       "      <td>Q28429317</td>\n",
       "      <td>Q729735</td>\n",
       "      <td>POSITIVE_DIAGNOSTIC_PREDICTOR_VpdD</td>\n",
       "      <td>JAK2 V617F</td>\n",
       "      <td>chronic myeloid leukemia</td>\n",
       "      <td>Sequence Variant</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32335</th>\n",
       "      <td>Q28371415</td>\n",
       "      <td>Q424401</td>\n",
       "      <td>VARIANT_OF_VvG</td>\n",
       "      <td>EGFR S492R</td>\n",
       "      <td>Epidermal growth factor receptor</td>\n",
       "      <td>Sequence Variant</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34863</th>\n",
       "      <td>Q30225387</td>\n",
       "      <td>Q23308049</td>\n",
       "      <td>HAS_PART_PWhpG</td>\n",
       "      <td>Folate Metabolism</td>\n",
       "      <td>Cystathionine-beta-synthase like</td>\n",
       "      <td>Biological Pathway</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59739</th>\n",
       "      <td>Q21139565</td>\n",
       "      <td>Q7868</td>\n",
       "      <td>CELL_COMPONENT_GccCC</td>\n",
       "      <td>Exportin-1</td>\n",
       "      <td>cell</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Cellular Component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159665</th>\n",
       "      <td>Q21108672</td>\n",
       "      <td>Q1057</td>\n",
       "      <td>BIOLOGICAL_PROCESS_GbpBP</td>\n",
       "      <td>Glycerol-3-phosphate acyltransferase, mitochon...</td>\n",
       "      <td>metabolism</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321210</th>\n",
       "      <td>Q21119889</td>\n",
       "      <td>Q14876109</td>\n",
       "      <td>MOLECULAR_FUNCTION_GmfMF</td>\n",
       "      <td>Kinesin family member 1A</td>\n",
       "      <td>motor activity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Molecular Function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408321</th>\n",
       "      <td>Q21123349</td>\n",
       "      <td>Q24721508</td>\n",
       "      <td>HAS_PART_GhpPD</td>\n",
       "      <td>Patched domain containing 1</td>\n",
       "      <td>Sterol-sensing domain</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Protein Domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438811</th>\n",
       "      <td>Q21096420</td>\n",
       "      <td>Q24719556</td>\n",
       "      <td>HAS_PART_GhpSS</td>\n",
       "      <td>Actin, gamma 2, smooth muscle, enteric</td>\n",
       "      <td>Actin/actin-like conserved site</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Super-Secondary Structure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442941</th>\n",
       "      <td>Q21105658</td>\n",
       "      <td>Q26916558</td>\n",
       "      <td>SUBCLASS_OF_GsPF</td>\n",
       "      <td>Cyclic nucleotide gated channel alpha 2</td>\n",
       "      <td>Cyclic nucleotide-gated channel alpha-2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Protein Family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442962</th>\n",
       "      <td>Q2720322</td>\n",
       "      <td>Q24738895</td>\n",
       "      <td>HAS_PART_GhpBS</td>\n",
       "      <td>Parvalbumin</td>\n",
       "      <td>EF-Hand 1, calcium-binding site</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Binding Site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443814</th>\n",
       "      <td>Q21119234</td>\n",
       "      <td>Q24783188</td>\n",
       "      <td>HAS_PART_GhpAS</td>\n",
       "      <td>Protein tyrosine phosphatase, receptor type U</td>\n",
       "      <td>Protein-tyrosine phosphatase, active site</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Active Site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444860</th>\n",
       "      <td>Q21115000</td>\n",
       "      <td>Q7948257</td>\n",
       "      <td>HAS_PART_GhpSM</td>\n",
       "      <td>Protein phosphatase 2 regulatory subunit Bbeta</td>\n",
       "      <td>WD40 repeat</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Structural Motif</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447098</th>\n",
       "      <td>Q29006010</td>\n",
       "      <td>Q123251</td>\n",
       "      <td>THERAPEUTIC_AREA_PPtaD</td>\n",
       "      <td>Introna</td>\n",
       "      <td>follicular lymphoma</td>\n",
       "      <td>Pharmaceutical Product</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448222</th>\n",
       "      <td>Q415173</td>\n",
       "      <td>Q12187</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>cefpodoxime</td>\n",
       "      <td>antibiotic</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458153</th>\n",
       "      <td>Q2840</td>\n",
       "      <td>Q7891</td>\n",
       "      <td>ANATOMICAL_LOCATION_DlA</td>\n",
       "      <td>influenza</td>\n",
       "      <td>respiratory system</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458857</th>\n",
       "      <td>Q21174786</td>\n",
       "      <td>Q7891</td>\n",
       "      <td>AFFLICTS_CHafA</td>\n",
       "      <td>Isophorone exposure</td>\n",
       "      <td>respiratory system</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460027</th>\n",
       "      <td>Q416410</td>\n",
       "      <td>Q21174292</td>\n",
       "      <td>TREATS_CtCH</td>\n",
       "      <td>fomepizole</td>\n",
       "      <td>ethylene glycol exposure</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         start_id     end_id                                 type  \\\n",
       "0         Q244408    Q407431    SIGNIFICANT_DRUG_INTERACTION_CdiC   \n",
       "1105      Q408801    Q170990                           TREATS_CtD   \n",
       "6322    Q47521989    Q409251          HAS_ACTIVE_INGREDIENT_PPaiC   \n",
       "8484    Q21175000    Q417316                      HAS_CAUSE_CHhcC   \n",
       "9081    Q44054558  Q27104149                       HAS_PART_PWhpC   \n",
       "12465    Q4847907    Q411917       PHYSICALLY_INTERACTS_WITH_GiwC   \n",
       "16110   Q28546459     Q18936  NEGATIVE_THERAPEUTIC_PREDICTOR_VntC   \n",
       "16651   Q29920046  Q27077182  POSITIVE_THERAPEUTIC_PREDICTOR_VptC   \n",
       "17372     Q844935   Q4650352              GENETIC_ASSOCIATION_DaG   \n",
       "20415    Q4454701    Q186889                        SYMPTOMS_DsyS   \n",
       "20787     Q373822      Q7867              MEDICAL_SPECIALTY_DmsMS   \n",
       "27320    Q7030575        Q86                       SYMPTOMS_CHsyS   \n",
       "30770   Q21167640   Q1799146                       SYMPTOMS_CHsyD   \n",
       "32251   Q27908043    Q264118   NEGATIVE_DIAGNOSTIC_PREDICTOR_VndD   \n",
       "32257   Q28429317    Q729735   POSITIVE_DIAGNOSTIC_PREDICTOR_VpdD   \n",
       "32335   Q28371415    Q424401                       VARIANT_OF_VvG   \n",
       "34863   Q30225387  Q23308049                       HAS_PART_PWhpG   \n",
       "59739   Q21139565      Q7868                 CELL_COMPONENT_GccCC   \n",
       "159665  Q21108672      Q1057             BIOLOGICAL_PROCESS_GbpBP   \n",
       "321210  Q21119889  Q14876109             MOLECULAR_FUNCTION_GmfMF   \n",
       "408321  Q21123349  Q24721508                       HAS_PART_GhpPD   \n",
       "438811  Q21096420  Q24719556                       HAS_PART_GhpSS   \n",
       "442941  Q21105658  Q26916558                     SUBCLASS_OF_GsPF   \n",
       "442962   Q2720322  Q24738895                       HAS_PART_GhpBS   \n",
       "443814  Q21119234  Q24783188                       HAS_PART_GhpAS   \n",
       "444860  Q21115000   Q7948257                       HAS_PART_GhpSM   \n",
       "447098  Q29006010    Q123251               THERAPEUTIC_AREA_PPtaD   \n",
       "448222    Q415173     Q12187                        HAS_ROLE_CrCR   \n",
       "458153      Q2840      Q7891              ANATOMICAL_LOCATION_DlA   \n",
       "458857  Q21174786      Q7891                       AFFLICTS_CHafA   \n",
       "460027    Q416410  Q21174292                          TREATS_CtCH   \n",
       "\n",
       "                                               start_name  \\\n",
       "0                                              diclofenac   \n",
       "1105                                            celecoxib   \n",
       "6322                                               Phoslo   \n",
       "8484                                 nitroethane exposure   \n",
       "9081                       Glycolysis and Gluconeogenesis   \n",
       "12465                   Cholinergic receptor muscarinic 4   \n",
       "16110                                  RB1 OVEREXPRESSION   \n",
       "16651                                       IDH2 MUTATION   \n",
       "17372                             coronary artery disease   \n",
       "20415                                  Taeniasis saginata   \n",
       "20787                                     eating disorder   \n",
       "27320                                  nicotine poisoning   \n",
       "30770                               acrylic acid exposure   \n",
       "32251                                         DNMT3A R882   \n",
       "32257                                          JAK2 V617F   \n",
       "32335                                          EGFR S492R   \n",
       "34863                                   Folate Metabolism   \n",
       "59739                                          Exportin-1   \n",
       "159665  Glycerol-3-phosphate acyltransferase, mitochon...   \n",
       "321210                           Kinesin family member 1A   \n",
       "408321                        Patched domain containing 1   \n",
       "438811             Actin, gamma 2, smooth muscle, enteric   \n",
       "442941            Cyclic nucleotide gated channel alpha 2   \n",
       "442962                                        Parvalbumin   \n",
       "443814      Protein tyrosine phosphatase, receptor type U   \n",
       "444860     Protein phosphatase 2 regulatory subunit Bbeta   \n",
       "447098                                            Introna   \n",
       "448222                                        cefpodoxime   \n",
       "458153                                          influenza   \n",
       "458857                                Isophorone exposure   \n",
       "460027                                         fomepizole   \n",
       "\n",
       "                                                 end_name  \\\n",
       "0                                                warfarin   \n",
       "1105                                            arthritis   \n",
       "6322                                      calcium acetate   \n",
       "8484                                          nitroethane   \n",
       "9081                                         D-malic acid   \n",
       "12465                                      glycopyrrolate   \n",
       "16110                                         doxorubicin   \n",
       "16651                                          Enasidenib   \n",
       "17372   ABO, alpha 1-3-N-acetylgalactosaminyltransfera...   \n",
       "20415                                              nausea   \n",
       "20787                                          psychiatry   \n",
       "27320                                            headache   \n",
       "30770                                 irritant dermatitis   \n",
       "32251                              acute myeloid leukemia   \n",
       "32257                            chronic myeloid leukemia   \n",
       "32335                    Epidermal growth factor receptor   \n",
       "34863                    Cystathionine-beta-synthase like   \n",
       "59739                                                cell   \n",
       "159665                                         metabolism   \n",
       "321210                                     motor activity   \n",
       "408321                              Sterol-sensing domain   \n",
       "438811                    Actin/actin-like conserved site   \n",
       "442941            Cyclic nucleotide-gated channel alpha-2   \n",
       "442962                    EF-Hand 1, calcium-binding site   \n",
       "443814          Protein-tyrosine phosphatase, active site   \n",
       "444860                                        WD40 repeat   \n",
       "447098                                follicular lymphoma   \n",
       "448222                                         antibiotic   \n",
       "458153                                 respiratory system   \n",
       "458857                                 respiratory system   \n",
       "460027                           ethylene glycol exposure   \n",
       "\n",
       "                   start_label                  end_label  \n",
       "0                     Compound                   Compound  \n",
       "1105                  Compound                    Disease  \n",
       "6322    Pharmaceutical Product                   Compound  \n",
       "8484           Chemical Hazard                   Compound  \n",
       "9081        Biological Pathway                   Compound  \n",
       "12465                     Gene                   Compound  \n",
       "16110         Sequence Variant                   Compound  \n",
       "16651         Sequence Variant                   Compound  \n",
       "17372                  Disease                       Gene  \n",
       "20415                  Disease                    Symptom  \n",
       "20787                  Disease          Medical Specialty  \n",
       "27320          Chemical Hazard                    Symptom  \n",
       "30770          Chemical Hazard                    Disease  \n",
       "32251         Sequence Variant                    Disease  \n",
       "32257         Sequence Variant                    Disease  \n",
       "32335         Sequence Variant                       Gene  \n",
       "34863       Biological Pathway                       Gene  \n",
       "59739                     Gene         Cellular Component  \n",
       "159665                    Gene         Biological Process  \n",
       "321210                    Gene         Molecular Function  \n",
       "408321                    Gene             Protein Domain  \n",
       "438811                    Gene  Super-Secondary Structure  \n",
       "442941                    Gene             Protein Family  \n",
       "442962                    Gene               Binding Site  \n",
       "443814                    Gene                Active Site  \n",
       "444860                    Gene           Structural Motif  \n",
       "447098  Pharmaceutical Product                    Disease  \n",
       "448222                Compound              Chemical Role  \n",
       "458153                 Disease       Anatomical Structure  \n",
       "458857         Chemical Hazard       Anatomical Structure  \n",
       "460027                Compound            Chemical Hazard  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# See a representitive edge of each type to ensure everything looks ok...\n",
    "combo.drop_duplicates(subset='type')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>start_name</th>\n",
       "      <th>end_name</th>\n",
       "      <th>start_label</th>\n",
       "      <th>end_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4573</th>\n",
       "      <td>Q127060</td>\n",
       "      <td>Q2070695</td>\n",
       "      <td>TREATS_CtD</td>\n",
       "      <td>folic acid</td>\n",
       "      <td>macrocytic anemia</td>\n",
       "      <td>Chemical Role</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20434</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q86</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>headache</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24048</th>\n",
       "      <td>Q7860879</td>\n",
       "      <td>Q101929</td>\n",
       "      <td>MEDICAL_SPECIALTY_DmsMS</td>\n",
       "      <td>type IV hypersensitivity</td>\n",
       "      <td>immunology</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>Medical Specialty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25153</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q2861470</td>\n",
       "      <td>MEDICAL_SPECIALTY_DmsMS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>emergency medicine</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Medical Specialty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27354</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q86</td>\n",
       "      <td>SYMPTOMS_CHsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>headache</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30775</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q1054718</td>\n",
       "      <td>SYMPTOMS_CHsyD</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>kidney disease</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63322</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>Q3824358</td>\n",
       "      <td>CELL_COMPONENT_GccCC</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Cellular Component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173242</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>Q14758909</td>\n",
       "      <td>BIOLOGICAL_PROCESS_GbpBP</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "      <td>G-protein coupled receptor signaling pathway</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323045</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>Q14819840</td>\n",
       "      <td>MOLECULAR_FUNCTION_GmfMF</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "      <td>neuropeptide hormone activity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Molecular Function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423721</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>Q24772514</td>\n",
       "      <td>HAS_PART_GhpPD</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "      <td>Glucagon/GIP/secretin/VIP</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Protein Domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451301</th>\n",
       "      <td>Q127060</td>\n",
       "      <td>Q183206</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>folic acid</td>\n",
       "      <td>vitamin B</td>\n",
       "      <td>Chemical Role</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454250</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>Q12537482</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "      <td>vasodilator agent</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459110</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q7891</td>\n",
       "      <td>AFFLICTS_CHafA</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>respiratory system</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2230</th>\n",
       "      <td>Q421239</td>\n",
       "      <td>Q408089</td>\n",
       "      <td>TREATS_CtD</td>\n",
       "      <td>D-penicillamine</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9098</th>\n",
       "      <td>Q44054573</td>\n",
       "      <td>Q127060</td>\n",
       "      <td>HAS_PART_PWhpC</td>\n",
       "      <td>One carbon metabolism and related pathways</td>\n",
       "      <td>folic acid</td>\n",
       "      <td>Biological Pathway</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35133</th>\n",
       "      <td>Q30225404</td>\n",
       "      <td>Q414964</td>\n",
       "      <td>HAS_PART_PWhpG</td>\n",
       "      <td>IL-6 signaling pathway</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "      <td>Biological Pathway</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164607</th>\n",
       "      <td>Q21101536</td>\n",
       "      <td>Q7860879</td>\n",
       "      <td>BIOLOGICAL_PROCESS_GbpBP</td>\n",
       "      <td>GATA binding protein 3</td>\n",
       "      <td>type IV hypersensitivity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449444</th>\n",
       "      <td>Q415220</td>\n",
       "      <td>Q127060</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>pemetrexed</td>\n",
       "      <td>folic acid</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460042</th>\n",
       "      <td>Q421239</td>\n",
       "      <td>Q408089</td>\n",
       "      <td>TREATS_CtCH</td>\n",
       "      <td>D-penicillamine</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         start_id     end_id                      type  \\\n",
       "4573      Q127060   Q2070695                TREATS_CtD   \n",
       "20434     Q408089        Q86             SYMPTOMS_DsyS   \n",
       "24048    Q7860879    Q101929   MEDICAL_SPECIALTY_DmsMS   \n",
       "25153     Q408089   Q2861470   MEDICAL_SPECIALTY_DmsMS   \n",
       "27354     Q408089        Q86            SYMPTOMS_CHsyS   \n",
       "30775     Q408089   Q1054718            SYMPTOMS_CHsyD   \n",
       "63322     Q414964   Q3824358      CELL_COMPONENT_GccCC   \n",
       "173242    Q414964  Q14758909  BIOLOGICAL_PROCESS_GbpBP   \n",
       "323045    Q414964  Q14819840  MOLECULAR_FUNCTION_GmfMF   \n",
       "423721    Q414964  Q24772514            HAS_PART_GhpPD   \n",
       "451301    Q127060    Q183206             HAS_ROLE_CrCR   \n",
       "454250    Q414964  Q12537482             HAS_ROLE_CrCR   \n",
       "459110    Q408089      Q7891            AFFLICTS_CHafA   \n",
       "2230      Q421239    Q408089                TREATS_CtD   \n",
       "9098    Q44054573    Q127060            HAS_PART_PWhpC   \n",
       "35133   Q30225404    Q414964            HAS_PART_PWhpG   \n",
       "164607  Q21101536   Q7860879  BIOLOGICAL_PROCESS_GbpBP   \n",
       "449444    Q415220    Q127060             HAS_ROLE_CrCR   \n",
       "460042    Q421239    Q408089               TREATS_CtCH   \n",
       "\n",
       "                                        start_name  \\\n",
       "4573                                    folic acid   \n",
       "20434                            mercury poisoning   \n",
       "24048                     type IV hypersensitivity   \n",
       "25153                            mercury poisoning   \n",
       "27354                            mercury poisoning   \n",
       "30775                            mercury poisoning   \n",
       "63322                Vasoactive intestinal peptide   \n",
       "173242               Vasoactive intestinal peptide   \n",
       "323045               Vasoactive intestinal peptide   \n",
       "423721               Vasoactive intestinal peptide   \n",
       "451301                                  folic acid   \n",
       "454250               Vasoactive intestinal peptide   \n",
       "459110                           mercury poisoning   \n",
       "2230                               D-penicillamine   \n",
       "9098    One carbon metabolism and related pathways   \n",
       "35133                       IL-6 signaling pathway   \n",
       "164607                      GATA binding protein 3   \n",
       "449444                                  pemetrexed   \n",
       "460042                             D-penicillamine   \n",
       "\n",
       "                                            end_name         start_label  \\\n",
       "4573                               macrocytic anemia       Chemical Role   \n",
       "20434                                       headache     Chemical Hazard   \n",
       "24048                                     immunology  Biological Process   \n",
       "25153                             emergency medicine     Chemical Hazard   \n",
       "27354                                       headache     Chemical Hazard   \n",
       "30775                                 kidney disease     Chemical Hazard   \n",
       "63322                                  intracellular                Gene   \n",
       "173242  G-protein coupled receptor signaling pathway                Gene   \n",
       "323045                 neuropeptide hormone activity                Gene   \n",
       "423721                     Glucagon/GIP/secretin/VIP                Gene   \n",
       "451301                                     vitamin B       Chemical Role   \n",
       "454250                             vasodilator agent                Gene   \n",
       "459110                            respiratory system     Chemical Hazard   \n",
       "2230                               mercury poisoning            Compound   \n",
       "9098                                      folic acid  Biological Pathway   \n",
       "35133                  Vasoactive intestinal peptide  Biological Pathway   \n",
       "164607                      type IV hypersensitivity                Gene   \n",
       "449444                                    folic acid            Compound   \n",
       "460042                             mercury poisoning            Compound   \n",
       "\n",
       "                   end_label  \n",
       "4573                 Disease  \n",
       "20434                Symptom  \n",
       "24048      Medical Specialty  \n",
       "25153      Medical Specialty  \n",
       "27354                Symptom  \n",
       "30775                Disease  \n",
       "63322     Cellular Component  \n",
       "173242    Biological Process  \n",
       "323045    Molecular Function  \n",
       "423721        Protein Domain  \n",
       "451301         Chemical Role  \n",
       "454250         Chemical Role  \n",
       "459110  Anatomical Structure  \n",
       "2230         Chemical Hazard  \n",
       "9098           Chemical Role  \n",
       "35133                   Gene  \n",
       "164607    Biological Process  \n",
       "449444         Chemical Role  \n",
       "460042       Chemical Hazard  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#take a look at some of these problematic nodes\n",
    "pd.concat([combo.query('start_id in @duplicated_nodes').drop_duplicates(subset=['start_id','type']),\n",
    "combo.query('end_id in @duplicated_nodes').drop_duplicates(subset=['end_id','type'])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# examining the above results, it looks like this is the mapping we desire\n",
    "resolved_multiples = {\n",
    "    'Q414964': 'Gene',\n",
    "    'Q127060': 'Compound', \n",
    "    'Q7860879': 'Biological Process',\n",
    "    'Q408089': 'Chemical Hazard'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_to_label = nodes.set_index('id')['label'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "nodes['label'] = nodes['id'].map(lambda i: resolved_multiples.get(i, id_to_label[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "66677\n",
      "66673\n"
     ]
    }
   ],
   "source": [
    "print(len(nodes))\n",
    "nodes = nodes.drop_duplicates().reset_index(drop=True)\n",
    "print(len(nodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "combo = gt.combine_nodes_and_edges(nodes, edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>start_name</th>\n",
       "      <th>end_name</th>\n",
       "      <th>start_label</th>\n",
       "      <th>end_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q244408</td>\n",
       "      <td>Q407431</td>\n",
       "      <td>SIGNIFICANT_DRUG_INTERACTION_CdiC</td>\n",
       "      <td>diclofenac</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q425295</td>\n",
       "      <td>Q7739</td>\n",
       "      <td>SIGNIFICANT_DRUG_INTERACTION_CdiC</td>\n",
       "      <td>cisapride</td>\n",
       "      <td>arsenic trioxide</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start_id   end_id                               type  start_name  \\\n",
       "0  Q244408  Q407431  SIGNIFICANT_DRUG_INTERACTION_CdiC  diclofenac   \n",
       "1  Q425295    Q7739  SIGNIFICANT_DRUG_INTERACTION_CdiC   cisapride   \n",
       "\n",
       "           end_name start_label end_label  \n",
       "0          warfarin    Compound  Compound  \n",
       "1  arsenic trioxide    Compound  Compound  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combo.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "combo['edge_abv'] = combo['type'].apply(lambda t: t.split('_')[-1])\n",
    "combo['actual_start'] = combo['edge_abv'].apply(lambda a: node_abv_to_full[gt.parse_edge_abbrev(a)[0]])\n",
    "combo['actual_end'] = combo['edge_abv'].apply(lambda a: node_abv_to_full[gt.parse_edge_abbrev(a)[2]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>start_name</th>\n",
       "      <th>end_name</th>\n",
       "      <th>start_label</th>\n",
       "      <th>end_label</th>\n",
       "      <th>edge_abv</th>\n",
       "      <th>actual_start</th>\n",
       "      <th>actual_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2230</th>\n",
       "      <td>Q421239</td>\n",
       "      <td>Q408089</td>\n",
       "      <td>TREATS_CtD</td>\n",
       "      <td>D-penicillamine</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>CtD</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20434</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q86</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>headache</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20490</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q209583</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>tachycardia</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20494</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q3002092</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>abdominal cramps</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20503</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q188008</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>dyspnea</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20591</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q223907</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>tremor</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20592</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q254327</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>anorexia</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20594</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q693058</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>chest pain</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20631</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q21117327</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>indecision</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20633</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q18343527</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>Gastrointestinal distress</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20638</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q21111972</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>eye irritation</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20667</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q9690</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>fatigue</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20725</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q35805</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>cough</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20742</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q2266426</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>visual disturbance</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20746</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q633403</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>weakness</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20750</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q623289</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>dysphagia</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>DsyS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24048</th>\n",
       "      <td>Q7860879</td>\n",
       "      <td>Q101929</td>\n",
       "      <td>MEDICAL_SPECIALTY_DmsMS</td>\n",
       "      <td>type IV hypersensitivity</td>\n",
       "      <td>immunology</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>Medical Specialty</td>\n",
       "      <td>DmsMS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Medical Specialty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25153</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>Q2861470</td>\n",
       "      <td>MEDICAL_SPECIALTY_DmsMS</td>\n",
       "      <td>mercury poisoning</td>\n",
       "      <td>emergency medicine</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Medical Specialty</td>\n",
       "      <td>DmsMS</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Medical Specialty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449444</th>\n",
       "      <td>Q415220</td>\n",
       "      <td>Q127060</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>pemetrexed</td>\n",
       "      <td>folic acid</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Compound</td>\n",
       "      <td>CrCR</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449493</th>\n",
       "      <td>Q421072</td>\n",
       "      <td>Q127060</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>pyrimethamine</td>\n",
       "      <td>folic acid</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Compound</td>\n",
       "      <td>CrCR</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449496</th>\n",
       "      <td>Q422232</td>\n",
       "      <td>Q127060</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>methotrexate</td>\n",
       "      <td>folic acid</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Compound</td>\n",
       "      <td>CrCR</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449498</th>\n",
       "      <td>Q422665</td>\n",
       "      <td>Q127060</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>trimethoprim</td>\n",
       "      <td>folic acid</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Compound</td>\n",
       "      <td>CrCR</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454250</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>Q12537482</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "      <td>vasodilator agent</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Chemical Role</td>\n",
       "      <td>CrCR</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455482</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>Q50377190</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "      <td>neuroprotective agent</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Chemical Role</td>\n",
       "      <td>CrCR</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455859</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>Q50377202</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "      <td>Gastrointestinal Agents</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Chemical Role</td>\n",
       "      <td>CrCR</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        start_id     end_id                     type  \\\n",
       "2230     Q421239    Q408089               TREATS_CtD   \n",
       "20434    Q408089        Q86            SYMPTOMS_DsyS   \n",
       "20490    Q408089    Q209583            SYMPTOMS_DsyS   \n",
       "20494    Q408089   Q3002092            SYMPTOMS_DsyS   \n",
       "20503    Q408089    Q188008            SYMPTOMS_DsyS   \n",
       "20591    Q408089    Q223907            SYMPTOMS_DsyS   \n",
       "20592    Q408089    Q254327            SYMPTOMS_DsyS   \n",
       "20594    Q408089    Q693058            SYMPTOMS_DsyS   \n",
       "20631    Q408089  Q21117327            SYMPTOMS_DsyS   \n",
       "20633    Q408089  Q18343527            SYMPTOMS_DsyS   \n",
       "20638    Q408089  Q21111972            SYMPTOMS_DsyS   \n",
       "20667    Q408089      Q9690            SYMPTOMS_DsyS   \n",
       "20725    Q408089     Q35805            SYMPTOMS_DsyS   \n",
       "20742    Q408089   Q2266426            SYMPTOMS_DsyS   \n",
       "20746    Q408089    Q633403            SYMPTOMS_DsyS   \n",
       "20750    Q408089    Q623289            SYMPTOMS_DsyS   \n",
       "24048   Q7860879    Q101929  MEDICAL_SPECIALTY_DmsMS   \n",
       "25153    Q408089   Q2861470  MEDICAL_SPECIALTY_DmsMS   \n",
       "449444   Q415220    Q127060            HAS_ROLE_CrCR   \n",
       "449493   Q421072    Q127060            HAS_ROLE_CrCR   \n",
       "449496   Q422232    Q127060            HAS_ROLE_CrCR   \n",
       "449498   Q422665    Q127060            HAS_ROLE_CrCR   \n",
       "454250   Q414964  Q12537482            HAS_ROLE_CrCR   \n",
       "455482   Q414964  Q50377190            HAS_ROLE_CrCR   \n",
       "455859   Q414964  Q50377202            HAS_ROLE_CrCR   \n",
       "\n",
       "                           start_name                   end_name  \\\n",
       "2230                  D-penicillamine          mercury poisoning   \n",
       "20434               mercury poisoning                   headache   \n",
       "20490               mercury poisoning                tachycardia   \n",
       "20494               mercury poisoning           abdominal cramps   \n",
       "20503               mercury poisoning                    dyspnea   \n",
       "20591               mercury poisoning                     tremor   \n",
       "20592               mercury poisoning                   anorexia   \n",
       "20594               mercury poisoning                 chest pain   \n",
       "20631               mercury poisoning                 indecision   \n",
       "20633               mercury poisoning  Gastrointestinal distress   \n",
       "20638               mercury poisoning             eye irritation   \n",
       "20667               mercury poisoning                    fatigue   \n",
       "20725               mercury poisoning                      cough   \n",
       "20742               mercury poisoning         visual disturbance   \n",
       "20746               mercury poisoning                   weakness   \n",
       "20750               mercury poisoning                  dysphagia   \n",
       "24048        type IV hypersensitivity                 immunology   \n",
       "25153               mercury poisoning         emergency medicine   \n",
       "449444                     pemetrexed                 folic acid   \n",
       "449493                  pyrimethamine                 folic acid   \n",
       "449496                   methotrexate                 folic acid   \n",
       "449498                   trimethoprim                 folic acid   \n",
       "454250  Vasoactive intestinal peptide          vasodilator agent   \n",
       "455482  Vasoactive intestinal peptide      neuroprotective agent   \n",
       "455859  Vasoactive intestinal peptide    Gastrointestinal Agents   \n",
       "\n",
       "               start_label          end_label edge_abv actual_start  \\\n",
       "2230              Compound    Chemical Hazard      CtD     Compound   \n",
       "20434      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20490      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20494      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20503      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20591      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20592      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20594      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20631      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20633      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20638      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20667      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20725      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20742      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20746      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "20750      Chemical Hazard            Symptom     DsyS      Disease   \n",
       "24048   Biological Process  Medical Specialty    DmsMS      Disease   \n",
       "25153      Chemical Hazard  Medical Specialty    DmsMS      Disease   \n",
       "449444            Compound           Compound     CrCR     Compound   \n",
       "449493            Compound           Compound     CrCR     Compound   \n",
       "449496            Compound           Compound     CrCR     Compound   \n",
       "449498            Compound           Compound     CrCR     Compound   \n",
       "454250                Gene      Chemical Role     CrCR     Compound   \n",
       "455482                Gene      Chemical Role     CrCR     Compound   \n",
       "455859                Gene      Chemical Role     CrCR     Compound   \n",
       "\n",
       "               actual_end  \n",
       "2230              Disease  \n",
       "20434             Symptom  \n",
       "20490             Symptom  \n",
       "20494             Symptom  \n",
       "20503             Symptom  \n",
       "20591             Symptom  \n",
       "20592             Symptom  \n",
       "20594             Symptom  \n",
       "20631             Symptom  \n",
       "20633             Symptom  \n",
       "20638             Symptom  \n",
       "20667             Symptom  \n",
       "20725             Symptom  \n",
       "20742             Symptom  \n",
       "20746             Symptom  \n",
       "20750             Symptom  \n",
       "24048   Medical Specialty  \n",
       "25153   Medical Specialty  \n",
       "449444      Chemical Role  \n",
       "449493      Chemical Role  \n",
       "449496      Chemical Role  \n",
       "449498      Chemical Role  \n",
       "454250      Chemical Role  \n",
       "455482      Chemical Role  \n",
       "455859      Chemical Role  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bad_edge = combo.query('start_label != actual_start or end_label != actual_end')\n",
    "bad_edge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Becasue combo is made by merging the nodes onto the edges, the index is constant between the two DataFrames\n",
    "edges = edges.drop(bad_edge.index).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write out results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "out_dir = '../2_pipeline/00_hetnet_design_and_build/out'\n",
    "\n",
    "if not os.path.exists(out_dir):\n",
    "    os.makedirs(out_dir)\n",
    "\n",
    "gt.add_colons(nodes).to_csv(os.path.join(out_dir, 'nodes.csv'), index=False)\n",
    "gt.add_colons(edges).to_csv(os.path.join(out_dir, 'edges.csv'), index=False)\n",
    "\n",
    "json.dump(node_info, open(os.path.join(out_dir, 'node_info.json'), 'w'), sort_keys=True, indent=2)\n",
    "json.dump(edge_info, open(os.path.join(out_dir, 'edge_info.json'), 'w'), sort_keys=True, indent=2)\n",
    "json.dump(hetnet_edges, open(os.path.join(out_dir, 'hetnet_edges.json'), 'w'), sort_keys=True, indent=2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
