{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "node_info = {}\n",
    "\n",
    "# There seem to be many types of sequence variants,\n",
    "# But the edges that link them are pretty unique, so will define based on\n",
    "# Those edges rather than a Q-ID\n",
    "node_info['Sequence Variant'] = {\n",
    "    'abbrev': 'V',\n",
    "    'id': None}\n",
    "\n",
    "node_info['Binding Site'] = {\n",
    "    'abbrev': 'BS',\n",
    "    'id': 'Q616005'}\n",
    "\n",
    "node_info['Active Site'] = {\n",
    "    'abbrev': 'AS',\n",
    "    'id': 'Q423026'}\n",
    "\n",
    "node_info['Super-Secondary Structure'] = {\n",
    "    'abbrev': 'SS',\n",
    "    'id': 'Q7644128'}\n",
    "\n",
    "node_info['Chemical Hazard'] = {\n",
    "    'abbrev': 'CH',\n",
    "    'id': 'Q21167512'}\n",
    "\n",
    "node_info['Molecular Function'] = {\n",
    "    'abbrev': 'MF',\n",
    "    'id': 'Q14860489'}\n",
    "    \n",
    "node_info['Pharmaceutical Product'] = {\n",
    "    'abbrev': 'PP',\n",
    "    'id': 'Q28885102'}\n",
    "\n",
    "node_info['Structural Motif'] = {\n",
    "    'abbrev': 'SM',\n",
    "    'id': 'Q3273544'}\n",
    "\n",
    "node_info['Protein Family'] = {\n",
    "    'abbrev': 'PF',\n",
    "    'id': 'Q417841'}\n",
    "\n",
    "node_info['Compound'] = {\n",
    "    'abbrev': 'C',\n",
    "    'id': 'Q11173'}\n",
    "\n",
    "node_info['Disease'] = {\n",
    "    'abbrev': 'D',\n",
    "    'id': 'Q12136'}\n",
    "\n",
    "## All proteins are encoded by genes, so qualify with 'gene encodes protein'\n",
    "node_info['Protein'] = {\n",
    "    'abbrev': 'P',\n",
    "    'id': 'Q8054',\n",
    "    'qualifiers': [{'s': 'Gene', 'p': 'P688', 'o': 'Protein', 'not': False},\n",
    "                   {'s': 'Gene', 'p': 'P703', 'o': 'Q15978631', 'not': False}]}\n",
    "\n",
    "## Genes will be defined as a human gene that does not encode a ANYTHING\n",
    "## None in Object means it only cares about P688.... (if you specifiy proteins, get memory erros...)\n",
    "node_info['Gene'] = {\n",
    "    'abbrev': 'G',\n",
    "    'id': 'Q7187',\n",
    "    'qualifiers': [{'s': 'Gene', 'p': 'P688', 'o': None, 'not': True}, \n",
    "                   {'s': 'Gene', 'p': 'P703', 'o': 'Q15978631', 'not': False}]}\n",
    "    \n",
    "node_info['Protein Domain'] = {\n",
    "    'abbrev': 'PD',\n",
    "    'id': 'Q898273'}\n",
    "    \n",
    "node_info['Cellular Component'] = {\n",
    "    'abbrev': 'CC',\n",
    "    'id': 'Q5058355'}\n",
    "    \n",
    "node_info['Biological Pathway'] = {\n",
    "    'abbrev': 'PW',\n",
    "    'id': 'Q4915012'}\n",
    "\n",
    "node_info['Biological Process'] = {\n",
    "    'abbrev': 'BP',\n",
    "    'id': 'Q2996394'}\n",
    "\n",
    "# Don't want symptoms that could also be diseases....\n",
    "# Get a lot of overlapping types espectially on chemcial hazard edges\n",
    "node_info['Symptom'] = {\n",
    "    'abbrev': 'S',\n",
    "    'id': 'Q169872',\n",
    "    'qualifiers': [{'s': 'Symptom', 'p': 'P31', 'o': 'Q12136', 'not': True},\n",
    "                   {'s': 'Symptom', 'p': 'P279', 'o': 'Q12136', 'not': True}]}\n",
    "    \n",
    "node_info['Medical Specialty'] = {\n",
    "    'abbrev': 'MS',\n",
    "    'id': 'Q930752'}\n",
    "\n",
    "# Chemical Role doesn't exist, but is an important type....\n",
    "# Again defining based on the edge\n",
    "node_info['Chemical Role'] = {\n",
    "    'abbrev': 'CR',\n",
    "    'id': None}\n",
    "    \n",
    "node_info['Anatomical Structure'] = {\n",
    "    'abbrev': 'A',\n",
    "    'id': 'Q4936952'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "edge_info = {}\n",
    "\n",
    "edge_info['BIOLOGICAL_PROCESS'] = {\n",
    "    'abbrev': 'bp',\n",
    "    'id': 'P682'}\n",
    "\n",
    "edge_info['BIOLOGICAL_VARIANT_OF'] = {\n",
    "    'abbrev': 'v',\n",
    "    'id': 'P3433'}\n",
    "\n",
    "edge_info['CELL_COMPONENT'] = {\n",
    "    'abbrev': 'cc',\n",
    "    'id': 'P681'}\n",
    "\n",
    "edge_info['MEDICAL_CONDITION_TREATED'] = {\n",
    "    'abbrev': 't',\n",
    "    'id': 'P2175',\n",
    "    'rev_id': 'P2176'}\n",
    "\n",
    "edge_info['GENETIC_ASSOCIATION'] = {\n",
    "    'abbrev': 'a',\n",
    "    'id': 'P2293'}\n",
    "\n",
    "edge_info['HAS_ACTIVE_INGREDIENT'] = {\n",
    "    'abbrev': 'ai',\n",
    "    'id': 'P3781'}\n",
    "\n",
    "edge_info['HAS_CAUSE'] = {\n",
    "    'abbrev': 'hc',\n",
    "    'id': 'P828'}\n",
    "\n",
    "edge_info['HAS_PART'] = {\n",
    "    'abbrev': 'hp',\n",
    "    'id': 'P527'}\n",
    "\n",
    "edge_info['MOLECULAR_FUNCTION'] = {\n",
    "    'abbrev': 'mf',\n",
    "    'id': 'P680'}\n",
    "\n",
    "edge_info['NEGATIVE_DIAGNOSTIC_PREDICTOR'] = {\n",
    "    'abbrev': 'nd',\n",
    "    'id': 'P3357'}\n",
    "\n",
    "edge_info['NEGATIVE_THERAPEUTIC_PREDICTOR'] = {\n",
    "    'abbrev': 'nt',\n",
    "    'id': 'P3355'}\n",
    "\n",
    "edge_info['PHYSICALLY_INTERACTS_WITH'] = {\n",
    "    'abbrev': 'iw',\n",
    "    'id': 'P129'}\n",
    "\n",
    "edge_info['POSITIVE_DIAGNOSTIC_PREDICTOR'] = {\n",
    "    'abbrev': 'pd',\n",
    "    'id': 'P3356'}\n",
    "\n",
    "edge_info['POSITIVE_THERAPEUTIC_PREDICTOR'] = {\n",
    "    'abbrev': 'pt',\n",
    "    'id': 'P3354'}\n",
    "\n",
    "edge_info['SUBCLASS_OF'] = {\n",
    "    'abbrev': 's',\n",
    "    'id': 'P279'}\n",
    "\n",
    "edge_info['SYMPTOMS'] = {\n",
    "    'abbrev': 'sy',\n",
    "    'id': 'P780'}\n",
    "\n",
    "edge_info['MEDICAL_SPECIALTY'] = {\n",
    "    'abbrev': 'ms',\n",
    "    'id': 'P1995'}\n",
    "\n",
    "edge_info['SIGNIFICANT_DRUG_INTERACTION'] = {\n",
    "    'abbrev': 'di',\n",
    "    'id': 'P769'}\n",
    "\n",
    "edge_info['AFFLICTS'] = {\n",
    "    'abbrev': 'af',\n",
    "    'id': 'P689'}\n",
    "\n",
    "edge_info['HAS_ROLE'] = {\n",
    "    'abbrev': 'r',\n",
    "    'id': 'P2868'}\n",
    "\n",
    "edge_info['THERAPEUTIC_AREA'] = {\n",
    "    'abbrev': 'ta',\n",
    "    'id': 'P4044'}\n",
    "\n",
    "edge_info['ANATOMICAL_LOCATION'] = {\n",
    "    'abbrev': 'l',\n",
    "    'id': 'P927'}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "node_abbreviations = {k: v['abbrev'] for k, v in node_info.items()}\n",
    "edge_abbreviations = {k: v['abbrev'] for k, v in edge_info.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "node_abv_to_full = {v: k for k, v in node_abbreviations.items()}\n",
    "edge_abv_to_full = {v: k for k, v in edge_abbreviations.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate queries for the edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../../hetnet-ml/src/')\n",
    "import graph_tools as gt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def generate_instance_tag(qname, n_id):\n",
    "    return \"\\n    {n} wdt:P31|wdt:P279 wd:{n_id} .\".format(n=qname, n_id=n_id)\n",
    "\n",
    "def to_query_name(name):\n",
    "    return '?'+name.lower().replace(' ', '_').replace('-', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_query_from_abbrev(abbrev, target=None):\n",
    "    # Split the edge abbreviation in to subject, predicate, object\n",
    "    s_abv, p_abv, o_abv = gt.parse_edge_abbrev(abbrev)\n",
    "    \n",
    "    # Get the full versions of names\n",
    "    s_name = node_abv_to_full[s_abv]\n",
    "    p_name = edge_abv_to_full[p_abv]\n",
    "    o_name = node_abv_to_full[o_abv]\n",
    "    \n",
    "    # Sanitize names for query purposes\n",
    "    s_qname = to_query_name(s_name)\n",
    "    o_qname = to_query_name(o_name)\n",
    "    \n",
    "    # Need to differentiate start and end if self referential edge\n",
    "    if s_qname == o_qname:\n",
    "        s_qname += '1'\n",
    "        o_qname += '2'\n",
    "    \n",
    "    # Get the proper return types\n",
    "    query_text = \"SELECT DISTINCT {s} {s}Label {o} {o}Label\".format(s=s_qname, o=o_qname)\n",
    "    query_text += \"\\nWHERE {\"\n",
    "    \n",
    "    return_types = {o_name}\n",
    "    \n",
    "    # Build the appropriate info to get the correct subject and object node-types\n",
    "    # Use set so not repeated in cases of self-referential edges\n",
    "    for name, qname in [(s_name, s_qname), (o_name, o_qname)]:\n",
    "        if node_info[name]['id']:\n",
    "            query_text += '\\n\\n    # Initial typing for {}'.format(name)\n",
    "            query_text += generate_instance_tag(qname, node_info[name]['id'])\n",
    "            return_types.update({name})\n",
    "    \n",
    "        # Get any qualifiers for subjects and objects\n",
    "        for qual in node_info[name].get('qualifiers', []):\n",
    "            query_text += '\\n    # Qualifier for {}'.format(name)\n",
    "            query_text += generate_qualifier(qual, return_types=return_types)\n",
    "            # Ensure there isn't the same nodetype definition multiple times\n",
    "            # E.G. ?gene wdt:P31|wdt:P279 wd:wd:Q7187 # ?gene subclass/instace of Gene\n",
    "            return_types.update({qual['s'], qual['o']})\n",
    "                \n",
    "    # Allows for changing of the target of an edge... for example if we want \n",
    "    # Sequnce Variant - variant of - Protein... SV's only go to Genes, so we can use \n",
    "    # SV - v - Gene - enc - Pro to get the relationship... \n",
    "    # Qualifiers in the protein definito handle the Gene-enc-Pro relationship, but the Target\n",
    "    # variable handles the fact that SVs point to genes\n",
    "    if target:\n",
    "        o_qname = to_query_name(target)\n",
    "    \n",
    "    # The actual edge we're interested in\n",
    "    query_text += \"\\n\\n    # Edge of interest {s} {p} {o}\".format(s=s_name, p=p_name, o=o_name)\n",
    "    \n",
    "    # Some edges are bi-directonal e.g. drug-used-for-treatment and medical-condition-treated\n",
    "    # Get edges in eitehr dircetion\n",
    "    rev_id = edge_info[p_name].get('rev_id', None)\n",
    "    if rev_id is not None:\n",
    "        query_text += \"\\n    {{ {s} wdt:{p} {o} }}\".format(s=s_qname, p=edge_info[p_name]['id'], o=o_qname)\n",
    "        query_text += \"\\n    UNION {{ {s} wdt:{p} {o} }}\".format(s=o_qname, p=rev_id, o=s_qname)\n",
    "    else:\n",
    "        query_text += \"\\n    {s} wdt:{p} {o} \".format(s=s_qname, p=edge_info[p_name]['id'], o=o_qname)\n",
    "\n",
    "    \n",
    "    # Make sure labels return in english\n",
    "    query_text += '\\n\\n    SERVICE wikibase:label { bd:serviceParam wikibase:language \"[AUTO_LANGAGE],en\" }\\n}'\n",
    "\n",
    "    return query_text\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def generate_qualifier(qual, return_types=None):\n",
    "    node_q_names = {}\n",
    "    \n",
    "    for node in ['s', 'o']:\n",
    "        node_name = qual[node]\n",
    "        if node_name in node_info.keys():\n",
    "            node_q_names[node] = to_query_name(node_name)\n",
    "        elif node_name is None:\n",
    "            node_q_names[node] = '?na'\n",
    "        else:    \n",
    "            node_q_names[node] = 'wd:'+node_name\n",
    "\n",
    "    qual_out = ''\n",
    "    \n",
    "    # if we have true node-types, we need to ensure they're used\n",
    "    for semtype, qname in node_q_names.items():\n",
    "        if not qname.startswith('wd:') and qual[semtype] not in return_types and qname is not '?na':\n",
    "            qual_out += generate_instance_tag(qname, node_info[qual[semtype]]['id'])        \n",
    "    \n",
    "    # Negative qualifiers need to filter out edges\n",
    "    if qual['not']:\n",
    "        qual_out += '\\n    FILTER NOT EXISTS {{ {s} wdt:{p} {o} }} .'.format(s=node_q_names['s'], \n",
    "                                                                            p=qual['p'], o=node_q_names['o'])\n",
    "    # Otherwise add the edges\n",
    "    else:\n",
    "        qual_out += '\\n    {s} wdt:{p} {o} .'.format(s=node_q_names['s'], \n",
    "                                                    p=qual['p'], o=node_q_names['o'])\n",
    "    \n",
    "    return qual_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_edge_info = [\n",
    "    {'abbrev': 'CdiC'},\n",
    "    {'abbrev': 'CtD'},\n",
    "    {'abbrev': 'PPaiC'},\n",
    "    {'abbrev': 'CHhcC'},\n",
    "    {'abbrev': 'PWhpC'},\n",
    "    {'abbrev': 'PiwC'},\n",
    "    {'abbrev': 'VntC'},\n",
    "    {'abbrev': 'VptC'},\n",
    "    {'abbrev': 'DaP', 'target': 'Gene'},\n",
    "    {'abbrev': 'DaG'},\n",
    "    {'abbrev': 'DsyS'},\n",
    "    {'abbrev': 'DmsMS'},\n",
    "    {'abbrev': 'CHsyS'},\n",
    "    {'abbrev': 'CHsyD'},\n",
    "    {'abbrev': 'VndD'},\n",
    "    {'abbrev': 'VpdD'},\n",
    "    {'abbrev': 'VvP', 'target': 'Gene'},\n",
    "    {'abbrev': 'VvG'},\n",
    "    {'abbrev': 'PWhpP', 'target': 'Gene'},\n",
    "    {'abbrev': 'PWhpG'},\n",
    "    {'abbrev': 'PccCC'},\n",
    "    {'abbrev': 'PbpBP'},\n",
    "    {'abbrev': 'PmfMF'},\n",
    "    {'abbrev': 'PhpPD'},\n",
    "    {'abbrev': 'PhpSS'},\n",
    "    {'abbrev': 'PsPF'},\n",
    "    {'abbrev': 'PhpBS'},\n",
    "    {'abbrev': 'PhpAS'},\n",
    "    {'abbrev': 'PhpSM'},\n",
    "    {'abbrev': 'PPtaD'},\n",
    "    {'abbrev': 'CrCR'},\n",
    "    {'abbrev': 'DlA'},\n",
    "    {'abbrev': 'CHafA'}\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "queries = [build_query_from_abbrev(**g_edge) for g_edge in graph_edge_info]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run the queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import functools\n",
    "from wikidataintegrator.wdi_core import WDItemEngine\n",
    "from tqdm import tqdm\n",
    "\n",
    "endpoint='http://avalanche.scripps.edu:9999/bigdata/sparql'\n",
    "\n",
    "def parse_result_uris(result):\n",
    "    for c in result:\n",
    "        if 'Label' not in c:\n",
    "            idx = result[c].str.startswith('http://www.wikidata.org/entity')\n",
    "            result[c][idx] = result[c][idx].str.split('/', expand=True).iloc[:,-1]\n",
    "    return result.drop_duplicates()\n",
    "\n",
    "query_func = functools.partial(WDItemEngine.execute_sparql_query, endpoint=endpoint, as_dataframe=True)\n",
    "\n",
    "def execute_sparql_query(query_text):\n",
    "    # Enforce the proper column order\n",
    "    col_order = query_text.split('\\n')[0].split(' ?')[1:]\n",
    "    return parse_result_uris(query_func(query_text))[col_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 33/33 [02:43<00:00,  1.66s/it]\n"
     ]
    }
   ],
   "source": [
    "res = {e['abbrev']: execute_sparql_query(q) for e, q in tqdm(zip(graph_edge_info, queries), total=len(graph_edge_info))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'460,735 raw edges'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many total edges retrieved?\n",
    "'{:,} raw edges'.format(sum(len(v) for v in res.values()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Two potnetial issues to look into...\n",
    "\n",
    "1. Self-referential edges...  \n",
    "    These edges may contain `node1 - node2` edges as well as `node2 - node1` edges\n",
    "    \n",
    "2. Multi or miss-typed nodes...  \n",
    "    Some node IDs may show up as multiple types... especially chemical roles and chemical compounds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_ids = dict()\n",
    "id_to_name = dict()\n",
    "self_ref = set()\n",
    "\n",
    "for e, r in res.items():\n",
    "    n_types = [c for c in r.columns if not c.endswith('Label')]\n",
    "    \n",
    "    for nt in n_types:\n",
    "        if nt.endswith('1') or nt.endswith('2'):\n",
    "            node_type = nt[:-1]\n",
    "            self_ref.add(e)\n",
    "        else:\n",
    "            node_type = nt\n",
    "    \n",
    "        if node_type in node_ids:\n",
    "            node_ids[node_type].update(set(r[nt]))\n",
    "        else:\n",
    "            node_ids[node_type] = set(r[nt])\n",
    "    \n",
    "        id_to_name.update(r.set_index(nt)[nt+'Label'].to_dict())\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fix the Self-Ref edges where forward and backward edges may be duplicated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(res.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_variant</th>\n",
       "      <th>sequence_variantLabel</th>\n",
       "      <th>protein</th>\n",
       "      <th>proteinLabel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q32965028</td>\n",
       "      <td>ABL1 BCR-ABL F311L</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q32965040</td>\n",
       "      <td>ABL1 BCR-ABL H396P</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q29938707</td>\n",
       "      <td>ABL1 BCR-ABL F359C</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q29938716</td>\n",
       "      <td>ABL1 BCR-ABL V299L</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q29938702</td>\n",
       "      <td>ABL1 BCR-ABL E255V</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q32964912</td>\n",
       "      <td>ABL1 E450G</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Q32965282</td>\n",
       "      <td>ABL1 M237V</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q32965302</td>\n",
       "      <td>ABL1 E281K</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q32965315</td>\n",
       "      <td>ABL1 A365V</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q32964949</td>\n",
       "      <td>ABL1 L298V</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Q32965063</td>\n",
       "      <td>ABL1 BCR-ABL P480A</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Q32965318</td>\n",
       "      <td>ABL1 A366G</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Q32965342</td>\n",
       "      <td>ABL1 I418V</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Q32964883</td>\n",
       "      <td>ABL1 BCR-ABL E355G</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Q32964888</td>\n",
       "      <td>ABL1 BCR-ABL L387M</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Q32964952</td>\n",
       "      <td>ABL1 BCR-ABL F311I</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Q32964959</td>\n",
       "      <td>ABL1 BCR-ABL E453Q</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Q32965015</td>\n",
       "      <td>ABL1 BCR-ABL F317V</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Q32965036</td>\n",
       "      <td>ABL1 BCR-ABL L364P</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Q32965044</td>\n",
       "      <td>ABL1 D421G</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Q32965299</td>\n",
       "      <td>ABL1 L273M</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Q32965345</td>\n",
       "      <td>ABL1 I418S</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Q32965019</td>\n",
       "      <td>ABL1 E279K</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Q32965023</td>\n",
       "      <td>ABL1 Q300R</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Q32965279</td>\n",
       "      <td>ABL1 Y232S</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Q32965288</td>\n",
       "      <td>ABL1 G250V</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Q32965313</td>\n",
       "      <td>ABL1 D363Y</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Q32965330</td>\n",
       "      <td>ABL1 Y393C</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Q29938714</td>\n",
       "      <td>ABL1 L384M</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Q32964908</td>\n",
       "      <td>ABL1 S438C</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>Q29938541</td>\n",
       "      <td>TP53 R280S</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>Q29938548</td>\n",
       "      <td>TP53 F270S</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>Q29938340</td>\n",
       "      <td>EGFR L747_S752del</td>\n",
       "      <td>Q424401</td>\n",
       "      <td>Epidermal growth factor receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>Q29938518</td>\n",
       "      <td>ABL1 BCR-ABL D276G</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>Q29938519</td>\n",
       "      <td>ABL1 F317L</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500</th>\n",
       "      <td>Q27908484</td>\n",
       "      <td>PDGFRA PDGFRA FUSIONS</td>\n",
       "      <td>Q21150807</td>\n",
       "      <td>Platelet-derived growth factor receptor alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2501</th>\n",
       "      <td>Q29938529</td>\n",
       "      <td>TP53 V272L</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2502</th>\n",
       "      <td>Q29938537</td>\n",
       "      <td>TP53 D281G</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2503</th>\n",
       "      <td>Q29938520</td>\n",
       "      <td>ABL1 M351T</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2504</th>\n",
       "      <td>Q29938381</td>\n",
       "      <td>TP53 R248L</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2505</th>\n",
       "      <td>Q29938530</td>\n",
       "      <td>TP53 G245D</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2506</th>\n",
       "      <td>Q29938539</td>\n",
       "      <td>TP53 D281H</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2507</th>\n",
       "      <td>Q29938509</td>\n",
       "      <td>ABL1 BCR-ABL M244V</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2508</th>\n",
       "      <td>Q29938514</td>\n",
       "      <td>ABL1 Q252H</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2509</th>\n",
       "      <td>Q29938383</td>\n",
       "      <td>TP53 R248FS</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2510</th>\n",
       "      <td>Q29938386</td>\n",
       "      <td>TP53 R213X</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2511</th>\n",
       "      <td>Q29938546</td>\n",
       "      <td>TP53 C275G</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2512</th>\n",
       "      <td>Q29938544</td>\n",
       "      <td>TP53 P278S</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513</th>\n",
       "      <td>Q29938516</td>\n",
       "      <td>ABL1 Y253F</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2514</th>\n",
       "      <td>Q27653851</td>\n",
       "      <td>ABL1 BCR-ABL</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2515</th>\n",
       "      <td>Q29938535</td>\n",
       "      <td>TP53 E285K</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2516</th>\n",
       "      <td>Q32965459</td>\n",
       "      <td>PXDNL P1460T</td>\n",
       "      <td>Q21145058</td>\n",
       "      <td>Peroxidasin-like protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2517</th>\n",
       "      <td>Q28445004</td>\n",
       "      <td>NT5E OVEREXPRESSION</td>\n",
       "      <td>Q21145399</td>\n",
       "      <td>5'-nucleotidase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2518</th>\n",
       "      <td>Q29938534</td>\n",
       "      <td>TP53 E286V</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2519</th>\n",
       "      <td>Q29938536</td>\n",
       "      <td>TP53 D281E</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2520</th>\n",
       "      <td>Q29938543</td>\n",
       "      <td>TP53 R280I</td>\n",
       "      <td>Q283350</td>\n",
       "      <td>Tumor protein p53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2521</th>\n",
       "      <td>Q28445054</td>\n",
       "      <td>RSF1 AMPLIFICATION</td>\n",
       "      <td>Q21151872</td>\n",
       "      <td>RSF1 protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2522</th>\n",
       "      <td>Q28371688</td>\n",
       "      <td>DDX41 R164W</td>\n",
       "      <td>Q21138065</td>\n",
       "      <td>Probable ATP-dependent RNA helicase DDX41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2523</th>\n",
       "      <td>Q28444986</td>\n",
       "      <td>CDK12 LOSS-OF-FUNCTION</td>\n",
       "      <td>Q21144067</td>\n",
       "      <td>Cyclin-dependent kinase 12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2524</th>\n",
       "      <td>Q28445200</td>\n",
       "      <td>BCOR MUTATION</td>\n",
       "      <td>Q21139279</td>\n",
       "      <td>BCL-6 corepressor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2525 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sequence_variant   sequence_variantLabel    protein  \\\n",
       "0           Q32965028      ABL1 BCR-ABL F311L    Q587961   \n",
       "1           Q32965040      ABL1 BCR-ABL H396P    Q587961   \n",
       "2           Q29938707      ABL1 BCR-ABL F359C    Q587961   \n",
       "3           Q29938716      ABL1 BCR-ABL V299L    Q587961   \n",
       "4           Q29938702      ABL1 BCR-ABL E255V    Q587961   \n",
       "5           Q32964912              ABL1 E450G    Q587961   \n",
       "6           Q32965282              ABL1 M237V    Q587961   \n",
       "7           Q32965302              ABL1 E281K    Q587961   \n",
       "8           Q32965315              ABL1 A365V    Q587961   \n",
       "9           Q32964949              ABL1 L298V    Q587961   \n",
       "10          Q32965063      ABL1 BCR-ABL P480A    Q587961   \n",
       "11          Q32965318              ABL1 A366G    Q587961   \n",
       "12          Q32965342              ABL1 I418V    Q587961   \n",
       "13          Q32964883      ABL1 BCR-ABL E355G    Q587961   \n",
       "14          Q32964888      ABL1 BCR-ABL L387M    Q587961   \n",
       "15          Q32964952      ABL1 BCR-ABL F311I    Q587961   \n",
       "16          Q32964959      ABL1 BCR-ABL E453Q    Q587961   \n",
       "17          Q32965015      ABL1 BCR-ABL F317V    Q587961   \n",
       "18          Q32965036      ABL1 BCR-ABL L364P    Q587961   \n",
       "19          Q32965044              ABL1 D421G    Q587961   \n",
       "20          Q32965299              ABL1 L273M    Q587961   \n",
       "21          Q32965345              ABL1 I418S    Q587961   \n",
       "22          Q32965019              ABL1 E279K    Q587961   \n",
       "23          Q32965023              ABL1 Q300R    Q587961   \n",
       "24          Q32965279              ABL1 Y232S    Q587961   \n",
       "25          Q32965288              ABL1 G250V    Q587961   \n",
       "26          Q32965313              ABL1 D363Y    Q587961   \n",
       "27          Q32965330              ABL1 Y393C    Q587961   \n",
       "28          Q29938714              ABL1 L384M    Q587961   \n",
       "29          Q32964908              ABL1 S438C    Q587961   \n",
       "...               ...                     ...        ...   \n",
       "2495        Q29938541              TP53 R280S    Q283350   \n",
       "2496        Q29938548              TP53 F270S    Q283350   \n",
       "2497        Q29938340       EGFR L747_S752del    Q424401   \n",
       "2498        Q29938518      ABL1 BCR-ABL D276G    Q587961   \n",
       "2499        Q29938519              ABL1 F317L    Q587961   \n",
       "2500        Q27908484   PDGFRA PDGFRA FUSIONS  Q21150807   \n",
       "2501        Q29938529              TP53 V272L    Q283350   \n",
       "2502        Q29938537              TP53 D281G    Q283350   \n",
       "2503        Q29938520              ABL1 M351T    Q587961   \n",
       "2504        Q29938381              TP53 R248L    Q283350   \n",
       "2505        Q29938530              TP53 G245D    Q283350   \n",
       "2506        Q29938539              TP53 D281H    Q283350   \n",
       "2507        Q29938509      ABL1 BCR-ABL M244V    Q587961   \n",
       "2508        Q29938514              ABL1 Q252H    Q587961   \n",
       "2509        Q29938383             TP53 R248FS    Q283350   \n",
       "2510        Q29938386              TP53 R213X    Q283350   \n",
       "2511        Q29938546              TP53 C275G    Q283350   \n",
       "2512        Q29938544              TP53 P278S    Q283350   \n",
       "2513        Q29938516              ABL1 Y253F    Q587961   \n",
       "2514        Q27653851            ABL1 BCR-ABL    Q587961   \n",
       "2515        Q29938535              TP53 E285K    Q283350   \n",
       "2516        Q32965459            PXDNL P1460T  Q21145058   \n",
       "2517        Q28445004     NT5E OVEREXPRESSION  Q21145399   \n",
       "2518        Q29938534              TP53 E286V    Q283350   \n",
       "2519        Q29938536              TP53 D281E    Q283350   \n",
       "2520        Q29938543              TP53 R280I    Q283350   \n",
       "2521        Q28445054      RSF1 AMPLIFICATION  Q21151872   \n",
       "2522        Q28371688             DDX41 R164W  Q21138065   \n",
       "2523        Q28444986  CDK12 LOSS-OF-FUNCTION  Q21144067   \n",
       "2524        Q28445200           BCOR MUTATION  Q21139279   \n",
       "\n",
       "                                           proteinLabel  \n",
       "0     ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "1     ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "2     ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "3     ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "4     ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "5     ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "6     ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "7     ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "8     ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "9     ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "10    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "11    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "12    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "13    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "14    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "15    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "16    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "17    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "18    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "19    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "20    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "21    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "22    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "23    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "24    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "25    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "26    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "27    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "28    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "29    ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "...                                                 ...  \n",
       "2495                                  Tumor protein p53  \n",
       "2496                                  Tumor protein p53  \n",
       "2497                   Epidermal growth factor receptor  \n",
       "2498  ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "2499  ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "2500      Platelet-derived growth factor receptor alpha  \n",
       "2501                                  Tumor protein p53  \n",
       "2502                                  Tumor protein p53  \n",
       "2503  ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "2504                                  Tumor protein p53  \n",
       "2505                                  Tumor protein p53  \n",
       "2506                                  Tumor protein p53  \n",
       "2507  ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "2508  ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "2509                                  Tumor protein p53  \n",
       "2510                                  Tumor protein p53  \n",
       "2511                                  Tumor protein p53  \n",
       "2512                                  Tumor protein p53  \n",
       "2513  ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "2514  ABL proto-oncogene 1, non-receptor tyrosine ki...  \n",
       "2515                                  Tumor protein p53  \n",
       "2516                           Peroxidasin-like protein  \n",
       "2517                                    5'-nucleotidase  \n",
       "2518                                  Tumor protein p53  \n",
       "2519                                  Tumor protein p53  \n",
       "2520                                  Tumor protein p53  \n",
       "2521                                       RSF1 protein  \n",
       "2522          Probable ATP-dependent RNA helicase DDX41  \n",
       "2523                         Cyclin-dependent kinase 12  \n",
       "2524                                  BCL-6 corepressor  \n",
       "\n",
       "[2525 rows x 4 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res['VvP']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00, 114.88it/s]\n"
     ]
    }
   ],
   "source": [
    "fixed = dict()\n",
    "\n",
    "for kind in tqdm(self_ref):\n",
    "    \n",
    "    # no need to worry about forward vs reverse in directed edges\n",
    "    if '>' in kind or '<' in kind:\n",
    "        continue\n",
    "    \n",
    "    # Only look at 1 kind of edge at a time\n",
    "    this_edges = res[kind]\n",
    "    col_names = this_edges.columns\n",
    "    \n",
    "    edge_ids = set()\n",
    "    \n",
    "    for row in this_edges.itertuples():\n",
    "        # Grab the edge ID, sorting, so lowest ID first:\n",
    "        #     If both 'Q00001 -- Q00002' and 'Q00002 -- Q00001' exist, effectively standarizes to \n",
    "        #     'Q00001 -- Q00002'\n",
    "        edge_id = tuple(sorted([row[1], row[3]]))\n",
    "        edge_ids.add(edge_id)\n",
    "\n",
    "    start_ids = []\n",
    "    start_names = []\n",
    "    end_ids = []\n",
    "    end_names = []\n",
    "\n",
    "        \n",
    "    for edge_id in edge_ids:\n",
    "        start_ids.append(edge_id[0])\n",
    "        start_names.append(id_to_name[edge_id[0]])\n",
    "        end_ids.append(edge_id[1])\n",
    "        end_names.append(id_to_name[edge_id[1]])\n",
    "    \n",
    "    fixed[kind] = pd.DataFrame({col_names[0]: start_ids, col_names[1]: start_names, col_names[2]: end_ids, col_names[3]: end_names})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1807, 1105)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(res['CdiC']), len(fixed['CdiC'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'66,677 nodes (possibly sume duplicats across type)'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# numver of unique node ids\n",
    "'{:,} nodes (possibly sume duplicats across type)'.format(sum(len(v) for v in node_ids.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "compound : 8087\n",
      "disease : 6205\n",
      "pharmaceutical_product : 2305\n",
      "chemical_hazard : 689\n",
      "biological_pathway : 563\n",
      "protein : 22742\n",
      "sequence_variant : 1758\n",
      "gene : 215\n",
      "symptom : 253\n",
      "medical_specialty : 69\n",
      "cellular_component : 1729\n",
      "biological_process : 12126\n",
      "molecular_function : 4267\n",
      "protein_domain : 4296\n",
      "supersecondary_structure : 455\n",
      "protein_family : 16\n",
      "binding_site : 55\n",
      "active_site : 89\n",
      "structural_motif : 108\n",
      "chemical_role : 494\n",
      "anatomical_structure : 156\n"
     ]
    }
   ],
   "source": [
    "for k, v in node_ids.items():\n",
    "    print('{} : {}'.format(k, len(v)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'66,673 unique nodes ids'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total = set()\n",
    "for v in node_ids.values():\n",
    "    total = total.union(v) \n",
    "'{:,} unique nodes ids'.format(len(total))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'4 total nodes with multiple types'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'{:,} total nodes with multiple types'.format(sum(len(v) for v in node_ids.values()) - len(total))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format nodes into Hetnet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes = []\n",
    "for k, v in node_ids.items():\n",
    "    curr_nodes = pd.DataFrame({'id': list(v), 'label': len(v)*[k]})\n",
    "    curr_nodes['name'] = curr_nodes['id'].map(id_to_name)\n",
    "    nodes.append(curr_nodes)\n",
    "nodes = pd.concat(nodes).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5545</th>\n",
       "      <td>Q127060</td>\n",
       "      <td>compound</td>\n",
       "      <td>folic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66483</th>\n",
       "      <td>Q127060</td>\n",
       "      <td>chemical_role</td>\n",
       "      <td>folic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12187</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>disease</td>\n",
       "      <td>mercury poisoning</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16826</th>\n",
       "      <td>Q408089</td>\n",
       "      <td>chemical_hazard</td>\n",
       "      <td>mercury poisoning</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1883</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>compound</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31845</th>\n",
       "      <td>Q414964</td>\n",
       "      <td>protein</td>\n",
       "      <td>Vasoactive intestinal peptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11420</th>\n",
       "      <td>Q7860879</td>\n",
       "      <td>disease</td>\n",
       "      <td>type IV hypersensitivity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47860</th>\n",
       "      <td>Q7860879</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>type IV hypersensitivity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             id               label                           name\n",
       "5545    Q127060            compound                     folic acid\n",
       "66483   Q127060       chemical_role                     folic acid\n",
       "12187   Q408089             disease              mercury poisoning\n",
       "16826   Q408089     chemical_hazard              mercury poisoning\n",
       "1883    Q414964            compound  Vasoactive intestinal peptide\n",
       "31845   Q414964             protein  Vasoactive intestinal peptide\n",
       "11420  Q7860879             disease       type IV hypersensitivity\n",
       "47860  Q7860879  biological_process       type IV hypersensitivity"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes[nodes.duplicated(keep=False, subset=['id'])].sort_values('id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "label_map = {to_query_name(k)[1:]: k for k in node_info.keys()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "nodes['label'] = nodes['label'].map(label_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q3429577</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Rhodamine 123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q474880</td>\n",
       "      <td>Compound</td>\n",
       "      <td>(21R)-argatroban</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q5282570</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Disperse Orange 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q27278051</td>\n",
       "      <td>Compound</td>\n",
       "      <td>transclopenthixol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q27114806</td>\n",
       "      <td>Compound</td>\n",
       "      <td>metamizole sodium hydrate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q5986473</td>\n",
       "      <td>Compound</td>\n",
       "      <td>icilin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Q368222</td>\n",
       "      <td>Compound</td>\n",
       "      <td>azapropazone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q3698301</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Cromakalim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q27108895</td>\n",
       "      <td>Compound</td>\n",
       "      <td>2-hydroxyglutarate(2-)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q5065694</td>\n",
       "      <td>Compound</td>\n",
       "      <td>3-[4-[[4-(aminomethyl)cyclohexyl]-oxomethoxy]p...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          id     label                                               name\n",
       "0   Q3429577  Compound                                      Rhodamine 123\n",
       "1    Q474880  Compound                                   (21R)-argatroban\n",
       "2   Q5282570  Compound                                  Disperse Orange 1\n",
       "3  Q27278051  Compound                                  transclopenthixol\n",
       "4  Q27114806  Compound                          metamizole sodium hydrate\n",
       "5   Q5986473  Compound                                             icilin\n",
       "6    Q368222  Compound                                       azapropazone\n",
       "7   Q3698301  Compound                                         Cromakalim\n",
       "8  Q27108895  Compound                             2-hydroxyglutarate(2-)\n",
       "9   Q5065694  Compound  3-[4-[[4-(aminomethyl)cyclohexyl]-oxomethoxy]p..."
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format Edges into Hetnet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges = []\n",
    "\n",
    "for k, v in res.items():\n",
    "    if k in fixed.keys():\n",
    "        v = fixed[k]\n",
    "    \n",
    "    \n",
    "    col_names = v.columns\n",
    "    keep_cols = [c for c in col_names if not c.endswith('Label')]\n",
    "    col_name_map = {keep_cols[0]: 'start_id', keep_cols[1]: 'end_id'}\n",
    "    \n",
    "    # Replace Proteins with Genes, to merge the protein and gene metanodes\n",
    "    parsed_edge = gt.parse_edge_abbrev(k)\n",
    "    if 'P' in parsed_edge:\n",
    "        idx = parsed_edge.index('P')\n",
    "        parsed_edge = list(parsed_edge)\n",
    "        parsed_edge[idx] = 'G'\n",
    "        k = ''.join(parsed_edge)\n",
    "    \n",
    "    v = v[keep_cols].rename(columns=col_name_map)\n",
    "    v['type'] = edge_abv_to_full[parsed_edge[1]] + '_' + k\n",
    "    \n",
    "    edges.append(v)\n",
    "edges = pd.concat(edges).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BIOLOGICAL_PROCESS_GbpBP               161545\n",
       "CELL_COMPONENT_GccCC                    99926\n",
       "MOLECULAR_FUNCTION_GmfMF                87111\n",
       "HAS_PART_GhpPD                          30490\n",
       "HAS_PART_PWhpG                          24882\n",
       "HAS_ROLE_CrCR                            9933\n",
       "MEDICAL_SPECIALTY_DmsMS                  6531\n",
       "MEDICAL_CONDITION_TREATED_CtD            5213\n",
       "HAS_PART_GhpSS                           4130\n",
       "PHYSICALLY_INTERACTS_WITH_GiwC           3645\n",
       "SYMPTOMS_CHsyS                           3450\n",
       "HAS_PART_PWhpC                           3384\n",
       "GENETIC_ASSOCIATION_DaG                  3043\n",
       "BIOLOGICAL_VARIANT_OF_VvG                2528\n",
       "HAS_PART_GhpSM                           2238\n",
       "HAS_ACTIVE_INGREDIENT_PPaiC              2166\n",
       "SYMPTOMS_CHsyD                           1481\n",
       "AFFLICTS_CHafA                           1170\n",
       "THERAPEUTIC_AREA_PPtaD                   1124\n",
       "SIGNIFICANT_DRUG_INTERACTION_CdiC        1105\n",
       "HAS_PART_GhpAS                           1046\n",
       "HAS_PART_GhpBS                            852\n",
       "POSITIVE_THERAPEUTIC_PREDICTOR_VptC       721\n",
       "ANATOMICAL_LOCATION_DlA                   704\n",
       "HAS_CAUSE_CHhcC                           597\n",
       "NEGATIVE_THERAPEUTIC_PREDICTOR_VntC       541\n",
       "SYMPTOMS_DsyS                             372\n",
       "POSITIVE_DIAGNOSTIC_PREDICTOR_VpdD         78\n",
       "SUBCLASS_OF_GsPF                           21\n",
       "NEGATIVE_DIAGNOSTIC_PREDICTOR_VndD          6\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges['type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# merge the genes and proteins in edges\n",
    "idx = nodes.query('label == \"Protein\"').index\n",
    "nodes.loc[idx, 'label'] = 'Gene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene                         22957\n",
       "Biological Process           12126\n",
       "Compound                      8087\n",
       "Disease                       6205\n",
       "Protein Domain                4296\n",
       "Molecular Function            4267\n",
       "Pharmaceutical Product        2305\n",
       "Sequence Variant              1758\n",
       "Cellular Component            1729\n",
       "Chemical Hazard                689\n",
       "Biological Pathway             563\n",
       "Chemical Role                  494\n",
       "Super-Secondary Structure      455\n",
       "Symptom                        253\n",
       "Anatomical Structure           156\n",
       "Structural Motif               108\n",
       "Active Site                     89\n",
       "Medical Specialty               69\n",
       "Binding Site                    55\n",
       "Protein Family                  16\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes['label'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "duplicated_nodes = nodes[nodes.duplicated(keep=False, subset=['id'])]['id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Q414964', 'Q127060', 'Q7860879', 'Q408089'], dtype=object)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicated_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "combo = gt.combine_nodes_and_edges(nodes, edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>start_name</th>\n",
       "      <th>end_name</th>\n",
       "      <th>start_label</th>\n",
       "      <th>end_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q18216</td>\n",
       "      <td>Q414275</td>\n",
       "      <td>SIGNIFICANT_DRUG_INTERACTION_CdiC</td>\n",
       "      <td>aspirin</td>\n",
       "      <td>tolbutamide</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>Q408535</td>\n",
       "      <td>Q131755</td>\n",
       "      <td>MEDICAL_CONDITION_TREATED_CtD</td>\n",
       "      <td>quetiapine</td>\n",
       "      <td>bipolar disorder</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6318</th>\n",
       "      <td>Q47521690</td>\n",
       "      <td>Q212272</td>\n",
       "      <td>HAS_ACTIVE_INGREDIENT_PPaiC</td>\n",
       "      <td>Hydrea</td>\n",
       "      <td>hydroxyurea</td>\n",
       "      <td>Pharmaceutical Product</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8484</th>\n",
       "      <td>Q21174183</td>\n",
       "      <td>Q416728</td>\n",
       "      <td>HAS_CAUSE_CHhcC</td>\n",
       "      <td>Diethylenetriamine exposure</td>\n",
       "      <td>diethylenetriamine</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9081</th>\n",
       "      <td>Q44015511</td>\n",
       "      <td>Q190012</td>\n",
       "      <td>HAS_PART_PWhpC</td>\n",
       "      <td>Methionine De Novo and Salvage Pathway</td>\n",
       "      <td>adenosine</td>\n",
       "      <td>Biological Pathway</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12465</th>\n",
       "      <td>Q21132915</td>\n",
       "      <td>Q265352</td>\n",
       "      <td>PHYSICALLY_INTERACTS_WITH_GiwC</td>\n",
       "      <td>Carbonic anhydrase 7</td>\n",
       "      <td>ethoxzolamide</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16110</th>\n",
       "      <td>Q28546459</td>\n",
       "      <td>Q18936</td>\n",
       "      <td>NEGATIVE_THERAPEUTIC_PREDICTOR_VntC</td>\n",
       "      <td>RB1 OVEREXPRESSION</td>\n",
       "      <td>doxorubicin</td>\n",
       "      <td>Sequence Variant</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16651</th>\n",
       "      <td>Q28531486</td>\n",
       "      <td>Q32089</td>\n",
       "      <td>POSITIVE_THERAPEUTIC_PREDICTOR_VptC</td>\n",
       "      <td>TSC1 FRAMESHIFT TRUNCATION</td>\n",
       "      <td>Sirolimus</td>\n",
       "      <td>Sequence Variant</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17372</th>\n",
       "      <td>Q1476525</td>\n",
       "      <td>Q21117791</td>\n",
       "      <td>GENETIC_ASSOCIATION_DaG</td>\n",
       "      <td>lipid metabolism disorder</td>\n",
       "      <td>Carnitine O-palmitoyltransferase 1, liver isoform</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20415</th>\n",
       "      <td>Q4454701</td>\n",
       "      <td>Q186889</td>\n",
       "      <td>SYMPTOMS_DsyS</td>\n",
       "      <td>Taeniasis saginata</td>\n",
       "      <td>nausea</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20787</th>\n",
       "      <td>Q373822</td>\n",
       "      <td>Q7867</td>\n",
       "      <td>MEDICAL_SPECIALTY_DmsMS</td>\n",
       "      <td>eating disorder</td>\n",
       "      <td>psychiatry</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Medical Specialty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27318</th>\n",
       "      <td>Q21173356</td>\n",
       "      <td>Q127076</td>\n",
       "      <td>SYMPTOMS_CHsyS</td>\n",
       "      <td>caprolactam exposure</td>\n",
       "      <td>vomiting</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Symptom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30768</th>\n",
       "      <td>Q21174753</td>\n",
       "      <td>Q19275831</td>\n",
       "      <td>SYMPTOMS_CHsyD</td>\n",
       "      <td>hydrogen chloride exposure</td>\n",
       "      <td>choking</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32249</th>\n",
       "      <td>Q27908043</td>\n",
       "      <td>Q264118</td>\n",
       "      <td>NEGATIVE_DIAGNOSTIC_PREDICTOR_VndD</td>\n",
       "      <td>DNMT3A R882</td>\n",
       "      <td>acute myeloid leukemia</td>\n",
       "      <td>Sequence Variant</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32255</th>\n",
       "      <td>Q28429317</td>\n",
       "      <td>Q729735</td>\n",
       "      <td>POSITIVE_DIAGNOSTIC_PREDICTOR_VpdD</td>\n",
       "      <td>JAK2 V617F</td>\n",
       "      <td>chronic myeloid leukemia</td>\n",
       "      <td>Sequence Variant</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32333</th>\n",
       "      <td>Q32965028</td>\n",
       "      <td>Q587961</td>\n",
       "      <td>BIOLOGICAL_VARIANT_OF_VvG</td>\n",
       "      <td>ABL1 BCR-ABL F311L</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "      <td>Sequence Variant</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34861</th>\n",
       "      <td>Q27436670</td>\n",
       "      <td>Q419337</td>\n",
       "      <td>HAS_PART_PWhpG</td>\n",
       "      <td>TCA Cycle (aka Krebs or citric acid cycle)</td>\n",
       "      <td>Dihydrolipoamide dehydrogenase</td>\n",
       "      <td>Biological Pathway</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59743</th>\n",
       "      <td>Q21119777</td>\n",
       "      <td>Q1308970</td>\n",
       "      <td>CELL_COMPONENT_GccCC</td>\n",
       "      <td>Heat shock protein family A (Hsp70) member 1B</td>\n",
       "      <td>inclusion body</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Cellular Component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159669</th>\n",
       "      <td>Q21106974</td>\n",
       "      <td>Q14633919</td>\n",
       "      <td>BIOLOGICAL_PROCESS_GbpBP</td>\n",
       "      <td>Promyelocytic leukemia</td>\n",
       "      <td>negative regulation of transcription, DNA-temp...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321214</th>\n",
       "      <td>Q21106037</td>\n",
       "      <td>Q14353133</td>\n",
       "      <td>MOLECULAR_FUNCTION_GmfMF</td>\n",
       "      <td>6-phosphofructo-2-kinase/fructose-2,6-biphosph...</td>\n",
       "      <td>catalytic activity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Molecular Function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408325</th>\n",
       "      <td>Q21144544</td>\n",
       "      <td>Q1829843</td>\n",
       "      <td>HAS_PART_GhpPD</td>\n",
       "      <td>Bromodomain-containing protein 9</td>\n",
       "      <td>Bromodomain</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Protein Domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438815</th>\n",
       "      <td>Q21117173</td>\n",
       "      <td>Q24779292</td>\n",
       "      <td>HAS_PART_GhpSS</td>\n",
       "      <td>Chromodomain Y-linked 2B</td>\n",
       "      <td>Chromo domain, conserved site</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Super-Secondary Structure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442945</th>\n",
       "      <td>Q21105658</td>\n",
       "      <td>Q26916558</td>\n",
       "      <td>SUBCLASS_OF_GsPF</td>\n",
       "      <td>Cyclic nucleotide gated channel alpha 2</td>\n",
       "      <td>Cyclic nucleotide-gated channel alpha-2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Protein Family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442966</th>\n",
       "      <td>Q21110951</td>\n",
       "      <td>Q24738895</td>\n",
       "      <td>HAS_PART_GhpBS</td>\n",
       "      <td>RAS guanyl releasing protein 1</td>\n",
       "      <td>EF-Hand 1, calcium-binding site</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Binding Site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443818</th>\n",
       "      <td>Q21119234</td>\n",
       "      <td>Q24783188</td>\n",
       "      <td>HAS_PART_GhpAS</td>\n",
       "      <td>Protein tyrosine phosphatase, receptor type U</td>\n",
       "      <td>Protein-tyrosine phosphatase, active site</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Active Site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444864</th>\n",
       "      <td>Q21115000</td>\n",
       "      <td>Q7948257</td>\n",
       "      <td>HAS_PART_GhpSM</td>\n",
       "      <td>Protein phosphatase 2 regulatory subunit Bbeta</td>\n",
       "      <td>WD40 repeat</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Structural Motif</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447102</th>\n",
       "      <td>Q29004950</td>\n",
       "      <td>Q189588</td>\n",
       "      <td>THERAPEUTIC_AREA_PPtaD</td>\n",
       "      <td>Capecitabine Teva</td>\n",
       "      <td>stomach cancer</td>\n",
       "      <td>Pharmaceutical Product</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448226</th>\n",
       "      <td>Q420354</td>\n",
       "      <td>Q12187</td>\n",
       "      <td>HAS_ROLE_CrCR</td>\n",
       "      <td>(-)-citrinin</td>\n",
       "      <td>antibiotic</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Chemical Role</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458159</th>\n",
       "      <td>Q2840</td>\n",
       "      <td>Q7891</td>\n",
       "      <td>ANATOMICAL_LOCATION_DlA</td>\n",
       "      <td>influenza</td>\n",
       "      <td>respiratory system</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458863</th>\n",
       "      <td>Q21174786</td>\n",
       "      <td>Q7891</td>\n",
       "      <td>AFFLICTS_CHafA</td>\n",
       "      <td>Isophorone exposure</td>\n",
       "      <td>respiratory system</td>\n",
       "      <td>Chemical Hazard</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         start_id     end_id                                 type  \\\n",
       "0          Q18216    Q414275    SIGNIFICANT_DRUG_INTERACTION_CdiC   \n",
       "1105      Q408535    Q131755        MEDICAL_CONDITION_TREATED_CtD   \n",
       "6318    Q47521690    Q212272          HAS_ACTIVE_INGREDIENT_PPaiC   \n",
       "8484    Q21174183    Q416728                      HAS_CAUSE_CHhcC   \n",
       "9081    Q44015511    Q190012                       HAS_PART_PWhpC   \n",
       "12465   Q21132915    Q265352       PHYSICALLY_INTERACTS_WITH_GiwC   \n",
       "16110   Q28546459     Q18936  NEGATIVE_THERAPEUTIC_PREDICTOR_VntC   \n",
       "16651   Q28531486     Q32089  POSITIVE_THERAPEUTIC_PREDICTOR_VptC   \n",
       "17372    Q1476525  Q21117791              GENETIC_ASSOCIATION_DaG   \n",
       "20415    Q4454701    Q186889                        SYMPTOMS_DsyS   \n",
       "20787     Q373822      Q7867              MEDICAL_SPECIALTY_DmsMS   \n",
       "27318   Q21173356    Q127076                       SYMPTOMS_CHsyS   \n",
       "30768   Q21174753  Q19275831                       SYMPTOMS_CHsyD   \n",
       "32249   Q27908043    Q264118   NEGATIVE_DIAGNOSTIC_PREDICTOR_VndD   \n",
       "32255   Q28429317    Q729735   POSITIVE_DIAGNOSTIC_PREDICTOR_VpdD   \n",
       "32333   Q32965028    Q587961            BIOLOGICAL_VARIANT_OF_VvG   \n",
       "34861   Q27436670    Q419337                       HAS_PART_PWhpG   \n",
       "59743   Q21119777   Q1308970                 CELL_COMPONENT_GccCC   \n",
       "159669  Q21106974  Q14633919             BIOLOGICAL_PROCESS_GbpBP   \n",
       "321214  Q21106037  Q14353133             MOLECULAR_FUNCTION_GmfMF   \n",
       "408325  Q21144544   Q1829843                       HAS_PART_GhpPD   \n",
       "438815  Q21117173  Q24779292                       HAS_PART_GhpSS   \n",
       "442945  Q21105658  Q26916558                     SUBCLASS_OF_GsPF   \n",
       "442966  Q21110951  Q24738895                       HAS_PART_GhpBS   \n",
       "443818  Q21119234  Q24783188                       HAS_PART_GhpAS   \n",
       "444864  Q21115000   Q7948257                       HAS_PART_GhpSM   \n",
       "447102  Q29004950    Q189588               THERAPEUTIC_AREA_PPtaD   \n",
       "448226    Q420354     Q12187                        HAS_ROLE_CrCR   \n",
       "458159      Q2840      Q7891              ANATOMICAL_LOCATION_DlA   \n",
       "458863  Q21174786      Q7891                       AFFLICTS_CHafA   \n",
       "\n",
       "                                               start_name  \\\n",
       "0                                                 aspirin   \n",
       "1105                                           quetiapine   \n",
       "6318                                               Hydrea   \n",
       "8484                          Diethylenetriamine exposure   \n",
       "9081               Methionine De Novo and Salvage Pathway   \n",
       "12465                                Carbonic anhydrase 7   \n",
       "16110                                  RB1 OVEREXPRESSION   \n",
       "16651                          TSC1 FRAMESHIFT TRUNCATION   \n",
       "17372                           lipid metabolism disorder   \n",
       "20415                                  Taeniasis saginata   \n",
       "20787                                     eating disorder   \n",
       "27318                                caprolactam exposure   \n",
       "30768                          hydrogen chloride exposure   \n",
       "32249                                         DNMT3A R882   \n",
       "32255                                          JAK2 V617F   \n",
       "32333                                  ABL1 BCR-ABL F311L   \n",
       "34861          TCA Cycle (aka Krebs or citric acid cycle)   \n",
       "59743       Heat shock protein family A (Hsp70) member 1B   \n",
       "159669                             Promyelocytic leukemia   \n",
       "321214  6-phosphofructo-2-kinase/fructose-2,6-biphosph...   \n",
       "408325                   Bromodomain-containing protein 9   \n",
       "438815                           Chromodomain Y-linked 2B   \n",
       "442945            Cyclic nucleotide gated channel alpha 2   \n",
       "442966                     RAS guanyl releasing protein 1   \n",
       "443818      Protein tyrosine phosphatase, receptor type U   \n",
       "444864     Protein phosphatase 2 regulatory subunit Bbeta   \n",
       "447102                                  Capecitabine Teva   \n",
       "448226                                       (-)-citrinin   \n",
       "458159                                          influenza   \n",
       "458863                                Isophorone exposure   \n",
       "\n",
       "                                                 end_name  \\\n",
       "0                                             tolbutamide   \n",
       "1105                                     bipolar disorder   \n",
       "6318                                          hydroxyurea   \n",
       "8484                                   diethylenetriamine   \n",
       "9081                                            adenosine   \n",
       "12465                                       ethoxzolamide   \n",
       "16110                                         doxorubicin   \n",
       "16651                                           Sirolimus   \n",
       "17372   Carnitine O-palmitoyltransferase 1, liver isoform   \n",
       "20415                                              nausea   \n",
       "20787                                          psychiatry   \n",
       "27318                                            vomiting   \n",
       "30768                                             choking   \n",
       "32249                              acute myeloid leukemia   \n",
       "32255                            chronic myeloid leukemia   \n",
       "32333   ABL proto-oncogene 1, non-receptor tyrosine ki...   \n",
       "34861                      Dihydrolipoamide dehydrogenase   \n",
       "59743                                      inclusion body   \n",
       "159669  negative regulation of transcription, DNA-temp...   \n",
       "321214                                 catalytic activity   \n",
       "408325                                        Bromodomain   \n",
       "438815                      Chromo domain, conserved site   \n",
       "442945            Cyclic nucleotide-gated channel alpha-2   \n",
       "442966                    EF-Hand 1, calcium-binding site   \n",
       "443818          Protein-tyrosine phosphatase, active site   \n",
       "444864                                        WD40 repeat   \n",
       "447102                                     stomach cancer   \n",
       "448226                                         antibiotic   \n",
       "458159                                 respiratory system   \n",
       "458863                                 respiratory system   \n",
       "\n",
       "                   start_label                  end_label  \n",
       "0                     Compound                   Compound  \n",
       "1105                  Compound                    Disease  \n",
       "6318    Pharmaceutical Product                   Compound  \n",
       "8484           Chemical Hazard                   Compound  \n",
       "9081        Biological Pathway                   Compound  \n",
       "12465                     Gene                   Compound  \n",
       "16110         Sequence Variant                   Compound  \n",
       "16651         Sequence Variant                   Compound  \n",
       "17372                  Disease                       Gene  \n",
       "20415                  Disease                    Symptom  \n",
       "20787                  Disease          Medical Specialty  \n",
       "27318          Chemical Hazard                    Symptom  \n",
       "30768          Chemical Hazard                    Disease  \n",
       "32249         Sequence Variant                    Disease  \n",
       "32255         Sequence Variant                    Disease  \n",
       "32333         Sequence Variant                       Gene  \n",
       "34861       Biological Pathway                       Gene  \n",
       "59743                     Gene         Cellular Component  \n",
       "159669                    Gene         Biological Process  \n",
       "321214                    Gene         Molecular Function  \n",
       "408325                    Gene             Protein Domain  \n",
       "438815                    Gene  Super-Secondary Structure  \n",
       "442945                    Gene             Protein Family  \n",
       "442966                    Gene               Binding Site  \n",
       "443818                    Gene                Active Site  \n",
       "444864                    Gene           Structural Motif  \n",
       "447102  Pharmaceutical Product                    Disease  \n",
       "448226                Compound              Chemical Role  \n",
       "458159                 Disease       Anatomical Structure  \n",
       "458863         Chemical Hazard       Anatomical Structure  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combo.drop_duplicates(subset='type')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "resolved_multiples = {\n",
    "    'Q414964': 'Gene',\n",
    "    'Q127060': 'Compound', \n",
    "    'Q7860879': 'Biological Process',\n",
    "    'Q408089': 'Chemical Hazard'\n",
    "}"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
